RFC for linking packages to their source and build (#626)
npm/rfcs: accepted/0000-link-packages-to-source-and-build.md

> # Link npm packages to the originating source code repository and build

Author: [View rendered version](https://github.com/npm/rfcs/blob/link-packages-to-source-and-build/accepted/0000-link-packages-to-source-and-build.md)

**Authors**: @feelepxyz @kommendorkapten @trevrosen

## Summary
Link public npm packages to the source code repository and build it originated from, making it harder to execute package hijacking attacks where a malicious version of an existing open source package gets uploaded to the registry.

Check out the [glossary](#glossary) for an overview of the different tools, terms and techniques covered in this RFC document.

## Motivation
So far our supply chain security offerings have focused on remediation - patching known vulnerabilities in dependencies (e.g. `npm audit`). The majority of vulnerabilities are accidental, so that was the right place to start. Now we need to turn our attention to deliberate supply chain attacks. In 2021, we saw these types of attacks increase by a terrifying [650%](https://www.sonatype.com/resources/state-of-the-software-supply-chain-2021).

> In November of last year alone, we saw three supply chain attacks against popular npm packages ([UA-Parser-JS](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack), [COA](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack), and [RC](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack)), each with millions of downloads per month.

Reviewer: ```suggestion
In November of last year alone, we saw three supply chain attacks against popular npm packages ([UAParser.js](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack), [Command-Option-Argument](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack), and [
](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack)), each with millions of downloads per month.
```

We should try to get the names of the projects as they self identify, rather than as what a third-party site (that also gets npm's name wrong) uses. These changes pulled with preference from the project's GitHub repo descriptions.

The following chart illustrates several attack vectors facing the supply chain:

> ![supply-chain-threats](https://user-images.githubusercontent.com/20165/174750708-2be483ac-7e41-4bc3-8ee9-440ef33d9423.svg)

Reviewer 2: I suggest citing this as being from SLSA.

Our focus is on mitigating package hijacking attacks where packages are built from a modified source (C) in the above diagram. This is where a malicious version of an existing open source package gets uploaded to the registry. These maliciously crafted versions are rarely reflected in the canonical source code repository and uploaded directly from an attacker's machine.

The [above](https://www.esentire.com/security-advisories/npm-library-supply-chain-attack) [attacks](https://www.esentire.com/security-advisories/coa-npm-supply-chain-attack) [are](https://www.esentire.com/security-advisories/rc-npm-supply-chain-attack) examples of this, and they frequently occur due to compromised npm credentials but can also happen due to compromised CI/CD or builds.

> Code on GitHub, GitLab etc can be browsed and audited, but packages in the registry are opaque and much harder to vet. If packages are built and published in the open these attacks become a whole lot harder. However, right now there’s no way of knowing where a package came from when you retrieve it from the registry.

Reviewer 3: unpkg.com makes it much easier - it seems like maybe addressing this (making npm packages browseable on the web safely) would be a good first step?

I'm reasonably certain that the owner of unpkg.com would be happy to transfer the IP to npm if it meant it'd be an official solution (at the least, it's worth asking)

Reviewer 4: We already have a beta implementation of exploring code in npm and are funding improvements to it for live code audit on npmjs.com in parallel to this work.

Reviewer 5: cc @feross

> We want to solve the problem of npm packages being disconnected from their source by linking the published package to the source code repository and build that it originated from.

Reviewer 3: given that build output should never be committed, and builds/publishes are typically done on local machines, how can this link be verified?

Reviewer 4: This RFC will only work with packages published from trusted build infrastructure. It will not support local publish.

Reviewer 3: What are the criteria for "trusted"? What are the associated costs in standing up such infrastructure, and what portions of the community would be excluded by those costs?

Reviewer 6: @ljharb I believe the premise of this entire RFC is predicated on builds/publishes happening in the cloud (as the motivation aims at getting closer to being able to meet SLSA Level 3 standards - ref. https://github.com/npm/rfcs/blob/link-packages-to-source-and-build/accepted/0000-link-packages-to-source-and-build.md#non-falsifiable-provenance-using-a-trusted-builder & https://slsa.dev/spec/v0.1/levels#detailed-explanation)

Reviewer 4: This is all covered in the RFC

https://github.com/npm/rfcs/pull/626/files#diff-95858aa7674c6ecb2578ad8c2b7ac848ac80dd36eedbbc92ce8a3968daf10771R239-R249

Reviewer 3: @darcyclarke without staged builds, I don't think that's something that the ecosystem will be able to adopt. Is that part of this RFC?

Author: > @darcyclarke without staged builds, I don't think that's something that the ecosystem will be able to adopt. Is that part of this RFC?

Staged builds are not currently part of this RFC but I can see how that could make it easier and safer to adopt automated publishes from ci/cd.

Reviewer 4: @ljharb I think we can use this RFC process to determine if your hypothesis is correct. Personally I do not believe that staged builds as a direct feature of npm will be necessary for ecosystem adoption and is not currently on our product roadmap.

It is possible to implement a ["manual approval" in GitHub actions](https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments) and I've seen customer evidence that [they are able to use this for npm today](https://twitter.com/meijer_s/status/1539535747212009474).

What is on the product roadmap are a number of improvements related to making publishing from CI more secure such as access tokens scoped to specific packages / scopes and streamlined setup experience to make it easier to get started.

Reviewer 3: If the manual approval allows me to input an actual second factor token, then that would be great! The things you linked don't seem to permit that yet tho.

I look forward to the RFC discussion.

Reviewer 4: @ljharb I'm surfacing that request to the actions team.

In general I see staging / approvals as out of scope for npm as a whole, but up for challenging those assumptions.

Reviewer 7: I would like to point out that the one of the major use cases of SLSA provenance is to mitigate the threat of compromised (Npm) registry credentials.  That's because the provenance provides a link to the source repo. So if an attacker were to compromise a registry credential and push a malicious package, the provenance would change from "source=github.com/genuine/source" to "source=github.com/attacker/source" or even "". The registry or another system that monitors it would immediately detect the compromise. In essence, this means that the system fails "safe" so long as:
1. you have 2FA on your registry account, protecting privileged settings, etc
1. you use an automation token to push packages.

An attacker who gets hold of an automation token can push packages but these would be immediately detected because of the change of source. In a nutshell, you don't really need staged builds if you upload provenance from a trusted builder.

Author: > An attacker who gets hold of an automation token can push packages but these would be immediately detected because of the change of source. In a nutshell, you don't need staged builds if you upload provenance from a trusted builder.

💯 Yes great summary! We can also make it difficult to change the provenance, e.g. locking it to a particular source repo once set, or preventing publishing new versions that don't include it if it's previously been set. Changing this could be gated behind 2FA on your registry account, limited to admins for an org.

Reviewer 8: If you disaggregate the system pushing the package to the repo from the system building/signing the artifact, you will be able to protect against the upload token being compromised because you can validate the signed artifacts with published public keys.

Furthermore, sigstore (I believe mentioned in this RFC, I'm still reading through it) can provide some ability to detect a compromised signing key. Will comment more on this later after reading more.

> Once a package includes provenance information (where and how it was built) we can start showing this when browsing a package on the npm registry. You would be able to click through to the particular commit and build that published a given version:

Reviewer: Ah, this was answered in the bullets below.

> Once a package includes provenance information (where and how it was built) we can start showing this when browsing a package on the npm registry. You would be able to click through to the particular commit and build that published a given version:

Author: > is it being designed in a way that will be proprietary to the npm website? Additionally, will the API be somewhat reproducible for those third-party registries for non-public modules?

We don't want it to be proprietary to the official npm registry, which is partly why we're excited about adopting Sigstore. We're planning to document what is required by third-party npm registries to support this functionality and make it as easy as possible to adopt.

> Once a package includes provenance information (where and how it was built) we can start showing this when browsing a package on the npm registry. You would be able to click through to the particular commit and build that published a given version:

Reviewer 8: There are several of us in both the CNCF and OpenSSF who are willing to help spec this out or provide guidance. Please let us know if you would like assistance.

<img width="376" alt="screenshot of npm package with provenance information" src="https://user-images.githubusercontent.com/20165/176637201-fdb02c11-c810-48e1-a203-ea4fb2008ad3.png">

> Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

Reviewer 3: Indeed; the entire point of `npm audit foo` when we discussed `npm audit signatures` was that every `foo` would be independently configurable and invokable.

> Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

Author: > existing `npm audit signatures` functionality and the new functionality as independent features

Thoughts on verifying all available signatures (registry and build) when running `npm audit signatures` but allowing you to filter the type with an argument like: `npm audit signatures --type=registry|build` (naming tbd)?

> Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

Reviewer 4: There is going to be broader shared behavior between the various signature verification commands such as "what do you do if packages are not yet in cache" that lead me to feeling a single command makes far more sense

> Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

Reviewer: FWIW I'm not at all against a single command but I have seen so many situations where people want *just one task* and don't care about other tasks in a bundled command.

It feels... very un-npm-y to me not to be able to opt-in to choosing more limited but precise sets of work that do the exact thing I want completed, especially when the functionality is fundamentally not related (checking what you received is what's on npm vs. what this RFC proposes).

> Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

Reviewer: Also to answer @feelepxyz: that's a totally fine solution to me!

> Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

Reviewer 3: `npm audit` is intended to be the single command - every subpart of it is intended to be granularly configurable and disableable by users, so that nobody is forced to do a check they don't want.

> Developers consuming open source packages should get the benefit of this without any changes to their workflows. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

Reviewer 8: If the consensus is to split it up into two separate functions, I would recommend splitting something like this:

```sh
npm audit signatures # runs everything related to signatures
npm audit signatures compare # only computes the hashes and compares
npm audit signatures integrity # validates the signature and ensures it is present and trusted in sigstore
```

This way, `npm audit` remains a single command, and `npm audit signatures` does not result in a user misunderstanding what it does.

> Open source maintainers should be able to add build provenance information to their packages with near-zero initial and ongoing overhead.

Reviewer: I'd like to recommend that `should be` should be changed to `must be`. I don't think shipping this with high friction that will be fixed later will result in a desirable outcome with regards to public perception/adoption of this functionality.

Good DX shouldn't be optional here, IMO.

Reviewer: ```suggestion
Open source maintainers must be able to add build provenance information to their packages with near-zero initial and ongoing overhead.
```

Author: Yes agreed!

> Open source maintainers must be able to add build provenance information to their packages with near-zero initial and ongoing overhead.

Reviewer 9: Given that local builds and publishes are not in-scope, this language should be toned down unless there's an "easy button" somewhere. 

If I'm understanding correctly, maintainers will have to change their publishing workflows to run in a "trusted" cloud environment. Maybe I'm out of the loop, but with TOTP enabled this has not historically been trivial to  set up. 

IMO this seems like a considerable barrier to adoption, at least today.

Author: @boneskull yes you're right, the main barrier will be changing publish workflows to run on supported CI/CD system. But beyond running the publish in a supported env, nothing else should need to be maintained or done, e.g. managing keys etc. I'll see how I can rephrase this. 

We're also planning on supporting publishing via OIDC in future which should make using CI/CD a lot more appealing. The setup for linking your source and build and publishing via OIDC would be identical.

So instead of using long-lived access tokens, or MFA/TOTP you could instead configure which repo/workflow is allowed to publish what package. The CLI would then exchange a OIDC id token from the CI/CD system for a short-lived access token which is used to publish the package.

Reviewer 4: It's also worth mentioning that automation tokens work today for publishing via CI when you have 2FA enabled on your accounts.

Author: Yes as part of this work we'll want to make sure it's easy to move publishing to CI/CD by providing a reusable workflow/trusted builder that supports common publish workflows. Hopefully good documentation and reusable publish workflows will alleviate some of the initial lift in moving publishing to CI/CD.

Reviewer 3: Automation tokens are a single factor, though.

Reviewer 4: Often when discussing multiple factors it is in the context of Authentication (the A in MFA). It is definitely important to have multiple factors used when creating a new session, and nothing about that is changed by utilizing an automation token. While it is a “single factor” credential, we have not historically seen ATO or project compromise due to access token theft, it has traditionally been from leaked passwords which were in turn then used to mint access tokens. The introduction of login verification and enforcement of 2FA (for authentication) of high-impact projects has all been focused on securing authentication primarily.

We do plan to continue to invest in making more secure practices for package publishing. Next quarter we will be kicking off a private beta for granular access tokens which will allow you to limit tokens to publish specific packages or  scopes, set expiration dates, and limit to specific CIDR ranges (only can publish from certain IPs). This will significantly improve the baseline security of using tokens in CI.

Longer term, we are exploring introducing OIDC as a publish mechanism which would allow publishing between a CI environment like actions and npm without the use of any long lived tokens. This would come coupled with experience improvements to make setting this process up on GitHub seamless, while laying an open foundation for other providers to implement similar integrations.

We’ve extensively reviewed these solutions with our security experts at GitHub and believe that they offer adequate security to projects, if they didn’t we would not make these features available. One can always be more secure, and I don’t say that to discourage folks from having excellent security hygiene (I have a yubikey on my keychain), but there does become a balance between usability and security and often a compromise is necessary. It is my personal opinion that the trade of of introducing a token and using CI to publish it worth the security benefits that we get knowing each publish is coming from a fresh machine.

## Goals

> - Establishes a verifiable link between a public npm package and the source repository and build it originated from.

Reviewer: It might be good to clarify at some point in this RFC if "source repository" must be GitHub or not. I'm 100% sure this will be a question that comes up, so it'd be good to be clear about that.

Author: There's no requirement on the source repo being a GitHub repo. The requirement is on the CI/CD system. So for example when CircleCI supports the Fulcio service (which verifies OIDC id tokens from the cicd system and mints short-lived signing certificates) you would be able to use any source repo supported by Circle.

Will clarify this 👍

Reviewer 8: Make sure to consider that the source git repo may change, e.g. tags are mutable. branches may be deleted and recreated with new content. Malicious content could be uploaded (by the dev, or through a compromised user credential), a build submitted, then the branch deleted and recreated in an attempt to hide the payload.

(May already be covered, still reading through the document).

Author: > Malicious content could be uploaded (by the dev, or through a compromised user credential), a build submitted, then the branch deleted and recreated in an attempt to hide the payload.

TBH we don't yet have great answers around these issues. We could try and mitigate against some things on the npm side, e.g. by optionally verifying the repo still exists during verification but verifying the correct code still exists is probably not feasible.

It would be great if we could somehow create immutable checkpoints in the source code repo/CI system. One naive idea could be to package up a shallow clone of the git repo as a build artifact 🤔

> - Does not expose any Personally identifiable information (PII) about maintainers, e.g. emails.

Reviewer 3: emails are already exposed for all publishers to npm - this seems like it should be a non-goal.

Reviewer 4: This is about the data that ends up in the public ledger, separate from the registry

Reviewer 3: ah ok, makes sense, thanks

Author: > emails are already exposed for all publishers to npm 

Yes this is true. We're not proposing changing what goes in the packument. The thing we've focused on in this RFC is the signed statement about the repo and build that will go on a public immutable signature ledger (Rekor). We didn't want a solution where this signed statement would include any kind of identifying information about maintainers.

Reviewer 8: I have a friend who is an expert and advocate on this exact topic in favor of maintaining developer privacy. Would it make sense to invite them to review the proposal and provide feedback?

Author: @fkautz yes please 👍  sounds like they might have some great context that we're also not aware of yet.

- Avoids developer-managed keys (as there's no good way to offer trust to the community given the challenges of distributing public keys). This, along with opt-in signing, adds near-zero initial and ongoing overhead for open source maintainers.

> - Maintainers can opt-in to including build provenance information (where the code lives and how it was built) when publishing using `npm publish` for public packages.

Reviewer: Describing in slightly more detail here what `build provenance information` is might be helpful.

- Incentivizes maintainers to build in the open because of the strong guarantees that this offers.
- Verification happens transparently on `npm install` without the need to obtain or manage additional tools or keys.
- Verification should have a negligible performance impact on `npm install`.
- Verification should be performed without depending on any third-party systems other than the registry.
- Compatible with third-party npm clients, e.g. `yarn` and `pnpm`.
- Should allow third-party npm registries, e.g. GitHub Packages, Artifactory and Verdaccio to follow suit and implement similar interfaces.

> - Should be maintained with >99.9% uptime so that developers are not blocked from publishing new packages.

Reviewer 3: 3 9's seems pretty paltry; can we aim higher?

Reviewer 2: Bear in mind this SLO affects publication only; it won't break builds if sigstore is offline. I expect most publishing authors will be fine with waiting a little bit.

Reviewer 10: +1, this SLO should be for signing which must be online (for fetching an identity certificate and publishing to the log).

Note that the verification mechanism should be designed with offline verification in mind. For example, querying the log for checking for inclusion would be an online action, but this can be designed to be offline with a persisted proof of inclusion.

Author: Yes this is only during publish when talking to Sigstore infrastructure. Sigstore is run as free-to-use public good infrastructure hosted on GCP which has SLAs from 99.9% to 99.99% on used services AFAICT. 

There's a lot to figure out committing to a SLOs on a shared service like this, with people on-call from different organisations. My hope would be that we can eventually get to four 9s uptime once this process is more mature.

Reviewer 8: I reached out to the sigstore team, and they pointed me to this as their current official statement, as of June 29, 2022: https://blog.sigstore.dev/an-update-on-general-availability-5c5563d4e400

They are targeting four months from June 29th (interpreting as end of October) to ramp up operations, tooling, and publishing SLOs.

Author: Yep I believe the current target for the Sigstore GA is 99.9% uptime for the hosted versions of Fulcio and Rekor.

- Should allow future extensions to support centrally managed signing authorities such as certificates managed by an enterprise and inner source within air-gapped enterprise environments.

> - Buy-in from the broader open source community.

Reviewer: Almost all of these seem to be effectively external, non-maintainer benefits. Are there any more maintainer-focused goals? Right now, I see very little incentive for maintainers to be excited for this outside of ones like me who work at companies and get paid to do this kind of maintenance.

Author: This is good feedback 👍  part of the story is not placing any kind of maintenance overhead on maintainers, for example having to manage keys. Another part in my mind is making publishing from CI/CD safer and easier, which is not covered in this RFC but we're planning as a follow up.

We want to allow authorising the publish using the same OIDC id token from the CI/CD system. So instead of using a long-lived access token to authorize the publish, you would log into your npm dashboard and set up which source repo is allowed to publish what package. The publish step then exchanges the OIDC id token for a short lived access token after the npm registry has checked the source repo from the id token matches the allowed one.

Once you've set up publishing in this way, you would not need to use long-lived access tokens and the risk of stolen credentials basically goes away if you also use 2FA on your npm account. The setup for this would be exactly the same as what we're proposing in this RFC so you essentially get it for free. I imagine adoption would go up a lot once we have both features in place and all major CI/CD providers are fully supported.

Reviewer 11: +1 to more clearly articulating the benefits to maintainers. You may also wish to separate goals (e.g. improve security) from requirements (e.g. zero overhead).

To me, the main benefit for maintainers is:

* Maintainers gain increased protection against malicious upload by compromised credentials or rogue maintainers.

- Support for self-hosted CI/CD systems.
- Mitigate every possible supply chain attack against the npm registry. For example:
  - Registry/mirror/proxy MITM attacks
  - Denial of service attacks against the registry or client from a compromised mirror/proxy
    - Rollback/fast-forward attacks (e.g. providing the wrong version for update)
    - Freeze/endless data attacks (e.g. preventing updates)
    - Extraneous dependencies attacks (e.g. tricking the client to install undeclared dependencies)
  - Malicious maintainer (e.g. [event-stream](https://blog.npmjs.org/post/180565383195/details-about-the-event-stream-incident))
  - Compromised login credentials (where 2FA can help)
  - Typosquatting attacks (e.g. `lodahs`), [“dependency confusion”](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610)
  - Accidental vulnerabilities

## Sigstore as signing solution
The [Sigstore](https://www.sigstore.dev/) project has been selected as the solution to signing npm packages (see detailed explanation below). It is currently the only working solution that supports our key requirements:

> - Links packages hosted on the npm registry to the source and build they originated from (provenance information).

Reviewer 3: other than recording "an npm specifier" and "a github repo and SHA" together, how are these linked? published npm packages already include repo and SHA information in the packument, so how would sigstore's linking be different?

Reviewer 4: This information would come as an OIDC claim from the publishing party and thus verifiable... e.g. the claim of a specific repo would come from the GitHub action itself, not meta data published to the registry from a personal machine

Reviewer 6: Just want to be mindful that there's nothing in this RFC which mentions that Actions/uild envs will become immutable which puts their audibility in question (ex. I can build & publish my package with a trusted builder but then delete that action/workflow - killing the link/trace)

Reviewer: Does this mean you _have_ to use GitHub as your source repo?

Reviewer 4: No

Author: We're proposing ways to make this information harder to falsify. Currently you can say the repo is anything and it's not validated, so an attacker can easily forge this information.

A part of the solution is using information from the OIDC id token that you can get from a supporting CI/CD system, this includes information about the repo and commit. The whole thing is then signed and can be verified against the CI/CD's public key for example.

Author: > nothing in this RFC which mentions that source, CI or build envs will be or will become immutable

Yes this is true. Worth noting though that the entry on Rekor and the stored copy on npm will show an audit trail of where it was originally built from. This might still be useful for post-mortem analysis. Having a broken link could also end being a strong signal to the community that a package is no longer maintained and should not be used.

> Does this mean you _have_ to use GitHub as your source repo?

No, we'll support any provider supported by Sigstore's Fulcio service. Today this only includes GitHub Actions but it's vendor neutral and any provider that meets the following requirements will be supported:
- OIDC ID tokens identifying the current workflow/run/build.
- ID token with a custom audience (needs to be set to `sigstore` for Fulcio to accept it).
- Claims about the code repository, commit, build and actor

Author: > I'm not sure if `npm` will want to do this but whether links are still valid is something that can be checked during package/artifact verification.

Yeah maybe this could be optional check that can be performed during verification. We probably don't want to do this by default as it could add significant time to the verification if we're checking 1000s+ links for a typical install.

Another idea would be to send the link though a proxy service on npm, and if this service detects a 404 it updates some state on the package that shows a big red warning next to the link next time you view the package on npmjs.com for example. This might cause more headaches than it's worth though, if requests intermittently fail or if the source repo is down and suddenly a bunch of links are broken and need to be fixed when the service is back up.

> - Supports signing identities that don't expose any personally identifiable information (PII) about maintainers, e.g. emails.

Reviewer 3: as noted above, emails are already public and exposed, so this isn't important

- Avoids developer-managed keys (as there's no good way to offer trust to the community given the challenges of distributing public keys).

Note that nothing in this current proposal precludes us from offering additional signing methods in the future as they become available and mature.

## Risks of adopting Sigstore
Sigstore infrastructure and tooling are currently not considered production ready and are run on a best-effort basis. The Sigstore project is working towards a [General Availability](https://blog.sigstore.dev/an-update-on-general-availability-5c5563d4e400) release later this year but we don't yet know what these guarantees will look like yet.

As such, there are several risks of adopting Sigstore for npm that are worth calling out:

> - No buy-in from the broader open source npm community.

Reviewer 3: tbh this seems like a potential dealbreaker

Reviewer: I don't disagree, but IMO this isn't because there wouldn't be but more because there's been near zero exposure.

- Sigstore's public and free-to-use infrastructure is not able to meet our uptime requirements (>99.9%).

> - Sigstore is not able to secure the funding required to run the infrastructure with production-grade support, SLOs and on-call rotation.

Reviewer 3: ```suggestion
- Sigstore is not able to secure the funding required to run the infrastructure with production-grade support, SLOs, and on-call rotation.
```

> To mitigate some of these risks, GitHub is planning to work directly with Sigstore to define and support robust production-grade SLOs for uptime and reliability that meet the requirements for npm. This will involve being on-call and wearing the pager for production services.

Reviewer: Is this a forever commitment, or is there a window in which GitHub will evaluate/re-evaluate this commitment? Additionally, is there a commitment from GitHub to messaging any changes around this if/when they do happen?

Author: I'm on the package security team at GitHub and we would be taking on this work. We're in the process if solidifying this commitment internally and working with the Sigstore org to figure out the details. We currently see this as an ongoing/foreseeable future commitment as long as npm/GitHub is relying on Sigstore services. We would be communicating any changes to this with long lead times.

Reviewer: Including some of this context in the RFC would (IMO) be super nice!

Author: Yes done 👍

In addition, GitHub is planning to support Sigstore by:

> - Running the signed timestamp authority for Rekor as well as a Certificate Transparency monitor for Fulcio. This will help spread the trust outside the four walls of Sigstore.

Reviewer 10: nit: Running `a` rather than `the`, since multiple timestamp authorities will be operated, one by Sigstore, and hopefully more by the community.

> - Running the signed timestamp authority for Rekor as well as a Certificate Transparency monitor for Fulcio. This will help spread the trust outside the four walls of Sigstore.

Reviewer 10: Ideally it's a monitor for both the Rekor transparency log and the Certificate Transparency log for Fulcio (The latter contains only certificates issued by Fulcio, while the former contains the artifact signatures)

- Maintaining the Sigstore trust root and having a root key holder as part of the group of Sigstore root key holders.

> We also remain open to an alternative solution if Sigstore is not able to meet npm's requirements for uptime and support.

Reviewer: Are there any potential alternatives that have been identified if this is the case?

Author: Not concrete alternatives but the rough shape would probably look similar to the architecture in this RFC with the difference that npm/github would be running the services instead. Either fully fledged or a paired down version.

It would be sad if we got to this as it would probably mean that we either end up with some proprietary signature service just for npm/github or potentially multiple competing public good instances.

Reviewer 8: Another alternative with minimal impact to the rest of this RFC is npm running or funding its own instance of Sigstore's software stack, primarily rekor.

## Detailed Explanation
### Sigstore as a solution to package signing
Package signing is a mitigation for package hijacking attacks. If maintainers sign their packages, end-users can verify that packages were produced by their corresponding owners. However, as of today, the main existing solution involves maintainer-managed PGP (or similar) keys, which has known [usability](https://www.usenix.org/legacy/events/sec99/full_papers/whitten/whitten.ps) [and](https://latacora.micro.blog/2019/07/16/the-pgp-problem.html) [security](https://blog.cryptographyengineering.com/2014/08/13/whats-matter-with-pgp/) [issues](https://www.wired.co.uk/article/efail-pgp-vulnerability-outlook-thunderbird-smime): registries that have implemented public key signing have seen either limited adoption or achieved adoption only by mandating it (e.g., Maven Central, which reports developer friction).

[Sigstore](https://www.sigstore.dev/) is a promising alternative: it allows signing artifacts with short-lived “disposable/ephemeral keys” notarized against log-ins with [OpenID Connect](https://openid.net/connect/) (OIDC) identities (e.g., Google or GitHub accounts).

> Several package ecosystems have announced interest in adopting Sigstore as a way for maintainers to sign packages (e.g., PyPI, RubyGems, Maven Central). [OpenSSF](https://blog.chainguard.dev/sigstore-statement-for-openssf-call/) is also working to standardize on Sigstore for signing software.

Reviewer 2: Suggest links:
* RubyGems: https://github.com/rubygems/rfcs/pull/37
* Maven: https://blog.sonatype.com/maven-central-and-sigstore
* PyPI: https://pypi.org/project/sigstore/

However, adopting something like Sigstore raises several major questions (beyond the risks detailed above):
- What security benefits does Sigstore provide?

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 2: There will inevitably be maintainers who object to signing for a variety of reasons. While I think it is reasonable to listen, a software repository also answers to the needs of end users.

For a maintainer a compromised package is perhaps embarrassing and distressing. For end users it is potentially catastrophic. And there are far more end users than maintainers. I don't see any other way to do the utilitarian calculus than to find that the needs of the end users for enhanced security are paramount.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 3: In any ecosystem where you have two cohorts of users, producers and consumers, it's simply not viable to do utilitarian calculus, because neither group can exist without the other. Both groups' needs must be weighed together.

eg, mandatory 2FA for maintainers is annoying and inconvenient, but the burden on maintainers is relatively small, and the benefits for maintainers and end users is exceedingly high, so it works out to enforce it.

If the burden is too large, or the benefits aren't bulletproof enough or high enough, it wouldn't likely be worth the tradeoffs.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 2: > Both groups' needs must be weighed together.

That is in fact the calculus. Two variables: cost to maintainers, cost to end users, integrated under a curve with respect to the spectrum of potential outcomes. The area under the curve is dominated by the impact on users because there are so many and because they are more negatively affected than maintainers by a compromise.

>If the burden is too large, or the benefits aren't bulletproof enough or high enough, it wouldn't likely be worth the tradeoffs.

The burden here is quite minimal. One click for the common case.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 3: I think it needs to be minimal for the uncommon cases too. Prolific maintainers can be eccentric, myself included, and doing both unpaid technical and emotional work does not tend to predispose humans to being charitable.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer: > For a maintainer a compromised package is perhaps embarrassing and distressing.

Sure, but this system doesn't meaningfully prevent this from happening. It _does_ prevent a certain version of this from happening, which IMO isn't necessarily a compelling story for me to add it outside of end-users maybe wanting this... but end-users want a lot of things, many of which are unreasonable/invalid. Social pressure shouldn't be the primary motivator.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer: > The burden here is quite minimal. One click for the common case.

Unfortunately, the most depended upon packages aren't the common case 😉

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Author: Copying my answer from another comment. Part of the story is not placing any kind of maintenance overhead on maintainers, for example not having to manage keys and making the UX/DX seamless. Saying that, the real burden with this proposal is moving publishing to CI/CD.

Arguably publishing from CI/CD today is actually less secure than from your laptop where you can use 2FA. We can do better.

We're also working on a proposal which will allow authorising the publish using the same OIDC id token from the CI/CD system. So instead of using a long-lived access token to authorize the publish, you would log into your npm dashboard and set up which source repo is allowed to publish what package. The publish step then exchanges the OIDC id token for a short lived access token after the npm registry has checked the source repo from the id token matches the allowed one.

Once you've set up publishing in this way, you would not need to use long-lived access tokens and the risk of stolen credentials basically goes away if you also use 2FA on your npm account. The setup for this would be exactly the same as what we're proposing in this RFC so you essentially get source and build linking for free if you set up OIDC publishing. I imagine adoption would pick up a lot once we have both features in place and all major CI/CD providers are fully supported.

We also want to provide reusable workflows that easy the burden of moving publishing to ci/cd.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 2: >Prolific maintainers can be eccentric, myself included, and doing both unpaid technical and emotional work does not tend to predispose humans to being charitable.

I understand and this dynamic is familiar to me as an observer of MFA rollouts in RubyGems and PyPI. My general point is that package maintainers, while an important constituency, are not the _only_ constituency. Software repositories owe a responsibility to all their stakeholders. Sometimes their interests and preferences will be in conflict and one constituency will necessarily be privileged over another, even though this will cause distress to some members of other constituencies.

This is such a case. I don't revel in the added work for package maintainers, however slight it is. But I see it as unavoidably necessary, given the world we now live in.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 2: >Social pressure shouldn't be the primary motivator.

I don't think it is. This is about cold hard numbers. The second most common attack on supply chains is account takeover. The third most common is repository compromise. Signing is a countermeasure to reduce the risk of both of these and is complementary to other countermeasures like MFA and implementing TUF.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 4: We are getting requests from maintainers and consumers that they want tools to make more informed and safe choices about the dependencies they consume. The solution that we've landed on for this is signed build attestations from trusted environments.

As this RFC outlines simply running `publish` from a trusted build environment (e.g. actions) would be all that's needed to do this. The surface DX for customers is something we will work together on fine tuning, but it mostly boils down to "is there a flag or not".

Here is an [example workflow](https://github.com/MylesBorins/node-osc/blob/main/.github/workflows/create-release.yml) that both publishes to npm and makes a GitHub release, if the RFC moves in the direction I want to advocate for my project would "just start signing" if I had a version of the npm CLI that supported signing. IMHO that is a fairly low barrier to entry.

Longer term we are looking at ways to significantly streamline the publishing process from GitHub -> npm built on open technologies such as OIDC that could make the process for developers setting up automations a "click of a button" with no managed tokens. While this is the long term goal I think it is out of scope for this RFC.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 3: Why is my local machine not capable of being a trusted build environment? A staggering amount of npm's downloads already trust it implicitly and have for almost a decade.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Author: > Why is my local machine not capable of being a trusted build environment?

With this proposal we want to figure out a way to say: this package on the npm registry, it was built from this source and you can go view it at this link to audit it if you like, and have some guarantees that it wasn't falsified. There's no way afaik to verifiably link the source repository to the published package from your local machine. You could, as a maintainer, say your package is from repo x, branch y and commit z but there's no way to verify this is actually legit information.

Taking this information from the OIDC id token generated by the CI/CD environment means it's not possible to falsify the information without the CI/CD systems private key used to sign it. With a reusable workflow that controls the build setup, we can also go one step further and make sure the repo and branch that you're building from is the same one that's specified in the OIDC id token, making sure the build hasn't gone and checked out some other branch to the one the action was triggered on for example.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 3: @feelepxyz a CI/CD machine is just as attackable as my local machine, though - perhaps more so, since it's a known target that's exposed to the open web. Why can't my local machine - something that has proven to be more trustworthy so far than all CI providers I'm aware of, largely due to nobody trying to attack it I'm sure - be able to sign it?

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Author: > @feelepxyz a CI/CD machine is just as attackable as my local machine, though - perhaps more so, since it's a known target that's exposed to the open web.

This is true and it definitely becomes a much more appealing target! 

> Why can't my local machine - something that has proven to be more trustworthy so far than all CI providers I'm aware of, largely due to nobody trying to attack it I'm sure - be able to sign it?

For me it isn't that I don't trust your personal build machine, its that we can't enforce any kind of provenance claims as to what repo the package was built from. An attacker that's publishing a malicious version of some package from their local machine could easily state that the package is from the original legitimate repo, without it actually reflecting the changes in the published package. 

What we could still do, and may want to enable this in future, is allow maintainers to sign from local machines using some identity that authenticates them as maintainers of the package, e.g. email based OIDC. This would then show *who* signed it but not *where* and *how*.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 12: > @feelepxyz a CI/CD machine is just as attackable as my local machine, though - perhaps more so, since it's a known target that's exposed to the open web. Why can't my local machine - something that has proven to be more trustworthy so far than all CI providers I'm aware of, largely due to nobody trying to attack it I'm sure - be able to sign it?

I would really argue that this isn't true, unless you're running something without any security controls. I would assume the CI/CD systems here would be follow standards, be pen tested, etc. Compare that with a local workstation that could be downloading arbitrary software, running arbitrary software in addition to the build you're running. That build could be attacked from a pretty unbounded number of vectors. Compare that with a CI/CD system that has adequate network, identity, etc. controls, there would be minimized attack surface.

> For me it isn't that I don't trust your personal build machine, its that we can't enforce any kind of provenance claims as to what repo the package was built from. An attacker that's publishing a malicious version of some package from their local machine could easily state that the package is from the original legitimate repo, without it actually reflecting the changes in the published package.

This doesn't fix the problem right? A CI/CD system that has been compromised can falsify some elements of the provenance just as well as a laptop.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 3: > This doesn't fix the problem right? A CI/CD system that has been compromised can falsify some elements of the provenance just as well as a laptop.

This is my point, I think - that there's nothing inherent about a CI/CD system that isn't also true about a personal laptop, and vice versa - the only difference is likelihoods due to competence, target awareness/attractiveness, which software is being run, etc.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 12: That's true. However, I think generally CI/CD is easier to secure compared to your laptop. CI/CD system can be audited, and configured through code, to ensure that for example that builds happen hermetically, only download pre-approved tools with a provenance trail. Your laptop unless it's managed by some other system doesn't haven't similar properties. Also in the case of packages that have multiple developers and maintainers it makes it more clear that packages are built through one system that has been approved by the project.

In the case of let's say GHA, you need to trust that GHA is doing the right things from a workflow perspective, OIDC, etc. GHA is taking on a lot of risk to their reputation, but most likely have resources to do so. If I build an npm package from my local laptop it's mostly on me. I could inadvertently have a rootkit on my machine, using a compromised OS kernel, have downloaded malicious software, etc. It is harder to trust anonymous actor's laptop compared to a potentially audited/threat modeled CI/CD system or service.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 3: Perhaps - but there are few personal laptop environments more objectively trustable than mine based on track record :-) without a doubt the majority will want to use this system, and probably should! However, if it's made a _requirement_ - explicitly, or implicitly via creating metadata that effectively punishes an author for not complying - that seems like it would be harming maintainers for a very dubious benefit.

> - How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.

Reviewer 2: Trust has economies of scale. Asking me to trust every random laptop means that, to achieve a similar level of confidence to a highly defended central provider, I would need to interview a bunch of people who, as has been pointed out elsewhere in this discussion, might object to the added burden.

Put another way: I don't know you. I know Github. Given my current priors, I trust them more than you, and it's costly for both you and I to update those priors.

#### What security benefits does Sigstore provide?
[Sigstore](https://www.sigstore.dev/) is a collection of tools and services aimed at making it easy to use short-lived, ephemeral certificates to sign software.

Sigstore has three main components: a CLI tool (Cosign), a Certificate Authority ([fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/)), and a time stamping and immutable ledger service ([rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/)).

> The Cosign functionality will be embedded directly in npm CLI, so we're left with the public [fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/) (CA) and [rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/) (ledger) services.

Reviewer: As someone unfamiliar with this ledger: Are there any concerns around exponential growth of this ledger and the insertion of new entries slowing down publishing? Other.... ledgers that I'm aware of begin to hit speed issues that would be concerning at the volume of _all published packages_.

Additionally, when you say `The Cosign functionality will be embedded directly in npm CLI` does this mean that the CLI will reach out to the Fulcio/Rekor APIs, or that additional binaries would be included?

Reviewer 10: The ledger is sharded - Every year a new tree is created, so this should alleviate concerns around performance. Additionally, while not currently implemented, we'd like to have log mirrors to distribute the load, but this will require the community operating mirrors, which comes with performance expectations and storage costs.

> The Cosign functionality will be embedded directly in npm CLI

I assume this is referring to performing the same behavior as Cosign, using the sigstore-js library to call out to Rekor/Fulcio.

Reviewer: > Every year a new tree is created, so this should alleviate concerns around performance.

I'd be really curious to know what the throughput is. npm has... a lot of publishes, and I get concerned about even a yearly reset if this will also include other ecosystems that are also growing + new ecosystems that haven't yet been invented.

What happens when the equivalent of all packages in 2022 are published in one week? 😅

Author: It will likely take a very long time before all daily package publishes are signing their builds. Assuming the uptick will be gradual over time.

The [database](https://github.com/google/trillian) that backs the transparency logs of both Rekor and Fulcio has very high throughput. It's the same database that powers the [certificate transparency log](https://transparency.dev/ ) for all SSL certs being issued and performs over 2,000 writes per second.

Reviewer: > It will likely take a very long time before all daily package publishes are signing their builds. Assuming the uptick will be gradual over time.

IIRC npm's growth over time hasn't necessarily been gradual 😅 

> The [database](https://github.com/google/trillian) that backs the transparency logs of both Rekor and Fulcio has very high throughput. It's the same database that powers the [certificate transparency log](https://transparency.dev/) for all SSL certs being issued and performs over 2,000 writes per second.

IMO it would be really nice to do some of the math before this RFC is ratified to see how much runway we have until we hit 2,000 writes per second across all the ecosystems that have begun considering this. It'd be good to know that this is/isn't a problem and how far out we need to start thinking about being concerned :)

Reviewer 2: I agree that some envelope arithmetic would be ideal. My current estimate is that it'll be no big deal: 2k RPS is 2.88M per day. From searching a bit I see counts of ~2M packages in npm. I doubt that every package is signed every day, so there's some headroom.

The 2k probably isn't a hard upper limit either. [Let's Encrypt reported early last year on their current DB servers](https://letsencrypt.org/2021/01/21/next-gen-database-servers.html) -- 2x EPYC giving 64 cores, with 2TB RAM. It's possible to buy [4 socket EPYC systems with up to 4TB of RAM](https://www.broadberry.com/superservers-supermicro-servers/sys-2014tp-htr).

##### What does [fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/) give us?

> - Independent party to validate claims of the OIDC identity token which contains references back to the repo, workflow run and git SHA.

Reviewer 10: If this RFC is only about provenance generated through CI, this is good.

Do you want to support other identities signing provenance or artifacts such as email or SPIFFE IDs?

Reviewer 13: For now, it's only focusing on CI systems integrated with Fulcio. I would say that it's harder to define policies around email and SPIFFE IDs. One central concept is to offer a set of "trusted builders" across all supported CI systems. As references to those should be part of the certificate (i.e they are validated by Fulcio) we have a way forward to identify/validate those packages built by a trusted builder.

But this does not of course rule out future extensions.

Author: > Do you want to support other identities signing provenance or artifacts such as email or SPIFFE IDs?

Possibly yes eventually! There's a whole question around private package and enterprise uses-cases that hasn't been answered here where we probably end up wanting to support several different signing identities.

Less sure about email identities as it really only tells us who published the package and not where and how the package was built. It also introduces the issue of linking that identity to an authorized maintainer. Open to exploring ideas around it though.

Reviewer 8: I'd like to help here if possible. I have experience with SPIFFE, SBOM generation, and in-toto and am happy to assist. I can also ask colleagues involved with projects like in-toto to help weigh in on the design.

The trick here I think will be to sign a top level document that embeds the claims. Perhaps the developer key can be used to sign the commit, then the commit can be embedded a document that is signed by the SPIFFE SVID, then submitted to Sigstore.

- A supported CI/CD provider can't falsify these claims and no one else can claim to be one of these CI/CD providers.
- The npm build script can't forge the source provenance by, for example, saying it's from some other repo or git SHA.
- Vendor-neutral: can support any standards-compliant OIDC Identity Provider that can meet their requirements.

##### What does [rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/) give us?
- Verifies that the signing happened during the time the certificate was valid and that the response from [Rekor](https://docs.sigstore.dev/rekor/overview/) is signed and can be verified offline.
- Verifiable, append-only ledger, makes it hard to tamper with already signed packages.
- Verification won't rely on [Rekor](https://docs.sigstore.dev/rekor/overview/) as we'll store detached signatures on the registry for performance and reliability but it adds an extra layer of protection in case of registry compromise.

#### How should maintainers authenticate to sign packages?
[Sigstore](https://www.sigstore.dev/) issues short-lived “disposable keys” notarized against log-ins with OIDC identities. So what OIDC identity provider (IdP) do maintainers use?

> This has been the subject of some controversy; for instance, in the [proposal to add Sigstore signing to RubyGems](https://github.com/rubygems/rfcs/pull/37), users worried about “vendorization” (making the repository reliant on third parties to function, e.g. GitHub, Google or Microsoft) and privacy (some maintainers are pseudonymous; Sigstore certificates include email addresses by default when signing using the [cosign](https://github.com/sigstore/cosign) CLI tool).

Reviewer 3: this should be considered a big concern for npm as well (not the email part tho)

Reviewer 8: Would npm consider being an OIDC Provider for users who do not want to rely on a third party provider? It already has unique user names, manages user identities, and has 2FA support.

Reviewer 10: It is an option, that was brought up in https://github.com/npm/rfcs/pull/626#discussion_r940621520.

While it does provide a good user experience, the concern is that it doesn't isolate the identity provider from the package repository, so a compromise of a user account for one would compromise both.

Sigstore supports both email and CI/CD workload-based OIDC identities:

##### Email based identity provider

> For instance, Google, GitHub, Microsoft (supported by Sigstore today) or any other IdP that Sigstore might support.

Reviewer: I know this is a problem introduced by proxy of using sigstore, but I'm uncomfortable with the likely future where maintainers who want to publish a package are relatively forced by their users to identify with a limited set of platforms that have extremely valid and well-documented criticisms around privacy (Google/Microsoft tracking, GitHub backing up on cookies policy),  politics (GitHub ICE, Microsoft DoD/JEDI deals, Google's Dragonfly, to name a few), and other ethical issues (treatment of women/Black/trans/other marginalized employees).

I don't fully believe that opting out of this feature for those reasons is a particularly equitable solution, nor one that will be an actual choice in the future.

Reviewer 2: This is mentioned as the Vendorization problem. [RubyGems was the first to encounter this discussion](https://github.com/rubygems/rfcs/pull/37#issuecomment-1024591418) and a number of alternatives were floated both within the RubyGems discussion and in discussion with peers in other ecosystems (including npm).

1. Grow the list of endorsed IdPs. The current list isn't meant to be final, it's just what was convenient to bootstrap. There are many candidate IdPs who are non-commercial.
2. Operate an IdP specific to npm. This solves the vendorization problem (you've already decided to trust npm) but it almost completely cancels a critical advantage of independent IdPs, which is that to sign and upload an attacker will need to compromise two accounts.
3. A "neutral" IdP, independent of npm, probably shared among multiple ecosystem software repositories, operated by some trusted third party (at a first guess, the OpenSSF).
4. An email-based flow for verification. This has mostly been rejected as "OIDC over SMTP" -- a real possibility of protocol design flaws make this highly risky.

Reviewer: I'm fine with #1 and maybe #3 (I fully understand why #2 and #4 are less than ideal). I'd really just want to see that list expanded somewhat significantly before this launches.

Author: We're sidestepping some of the problems with email based identities in this proposal by just supporting workflow identities from supported CI/CD systems. That said, we might wan to support email based identities in future if support improves.

Reviewer 14: I certainly hope vendorization will be addressed from the beginning by offering alternative solutions. As other package managers run into the same issue, I see an opportunity for collaboration. If vendorization isn't addressed, you can expect a vocal group to resent this proposal, causing confusion and doubt and ultimately leading to failure of the whole thing. Improving supply chain security is too important to fail, so it should be done right to get everybody on board.

Reviewer 2: For reference, [some further discussion of](https://github.com/sigstore/fulcio/issues/444) the Neutral IdP idea took place in Fulcio issue 444.

Author: > I certainly hope vendorization will be addressed from the beginning by offering alternative solutions.

💯 One of our internal goals of publishing this RFC is to help start these conversions with CI providers. I'm also excited about Sigstore as a solution to software signing for this reason. It's a vendor neutral entity that can help create alignment.

I'm not closed or opposed to extending npm package signing to support some kind of "author/maintainer" IdP in future that could allow maintainers to publish from their laptops. For the first iteration I think we'll start with only supporting build or workflow identities from CI/CD providers.

**Pros**
- Separate party for trust: it would be difficult for an attacker to compromise the repository and one of these IdPs.
- Many (but not all!) users have these accounts already.
- Portable: supports publishing from most environments (requiring a browser and an IdP login).
- Minimal changes to existing publish workflow for maintainers that are publishing directly.
- Reliability: these IdPs already work at scale and have full-time support.

**Cons**
- Privacy: the identities used to sign packages are public (though users can make a pseudonymous account).
- Privacy: Only a handful of identity providers are supported today that might not suit all users (GitHub, Google or Microsoft).

> - Security: Some users might only have secondary, less secure, accounts on one of the supported services making it an easier target.

Reviewer 10: Agreed, no way to check that a user has MFA enabled on one of those accounts for example, as this isn't included in identity tokens.

Reviewer: IMO this should borderline be a requirement. I'm amazed that this isn't supported. If my publishing email is `j@gmail.com` but I publish the password publicly, this means anyone could sign in to that email and publish as me which... fundamentally gets around the entire point of this.

Reviewer 10: A concern that has been brought up in the past is including the MFA status in identity tokens makes it easier to find targets. It would probably be simpler for all identity providers to require MFA.

Reviewer: > It would probably be simpler for all identity providers to require MFA.

No opposition from me :)

- The email might not match the one used on the registry making it hard to validate that it's from a legitimate maintainer.

##### CI/CD workload identity provider
For instance, Github Actions (supported by Sigstore today) or any other CI/CD system that can provide OIDC workload identities with required claims.

**Pros**
- Support for pseudonyms/privacy: does not require public identities other than what already corresponds to the package (e.g. repository).
- Links the package to the source repository and includes information about and how it was built: commit SHA, workflow run id etc.
- Vendor-neutral: supports any CI/CD system supporting OIDC workload identities.
- OIDC workflow identities could also be used to authenticate the package publish (restricting publishing to a specific source code repo and removing the need for long-lived access tokens).

**Cons**
- Limited support today: Only GitHub Actions is fully supported in Fulcio today.
- Friction: Requires publishing from supported CI/CD providers.
- CI/CD providers will need to operate an IdP and get it integrated with Sigstore.

> Using workload identities along with short-lived x.509 certificates is currently the only solution that meets our key requirements:

Reviewer 10: nit: X.509

- Linking packages to the source and build.
- Doesn't expose any personal identifiable information about maintainers.
- Avoids developer-managed keys by placing trust in the CI/CD identity provider.

> Today this means only commercial CI/CD providers will be supported. This is far from where all developers are publishing today so this will add friction and slow adoption. All major CI/CD providers offer free plans for Open Source projects and we hope that all packages will eventually be built out in the open to make supply chain attacks harder to execute.

Reviewer: If this is the goal, you are going to get questions about 2FA and automated publishing. While strides have been made, are there further efforts to continue to make publishing from CI a reasonable experience given that this will effectively force _all_ maintainers of large-scale packages to publish from CI?

Author: Copying answer from another thread. 

> are there further efforts to continue to make publishing from CI a reasonable experience given that this will effectively force all maintainers of large-scale packages to publish from CI?

Yes! We're working on a proposal which will allow authorising the publish using the same OIDC id token from the CI/CD system. So instead of using a long-lived access token to authorize the publish, you would log into your npm dashboard and set up which source repo is allowed to publish what package. The publish step then exchanges the OIDC id token for a short lived access token after the npm registry has checked the source repo from the id token matches the allowed one.

Once you've set up publishing in this way, you would not need to use long-lived access tokens and the risk of stolen credentials basically goes away if you also use 2FA on your npm account. The setup for this would be exactly the same as what we're proposing in this RFC so you essentially get source and build linking for free if you set up OIDC publishing. I imagine adoption would pick up a lot once we have both features in place and all major CI/CD providers are fully supported.

I think we can also ease the migration by providing reusable workflows for publishing npm packages on CI/CD. On actions you could start by using a [manual workflow trigger](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/) to publish.

Reviewer 4: While longer term we are hoping to implement something on top of OIDC we are currently working on Granular access tokens as a successor to automation tokens. These tokens will:

* Be able to be scoped to a specific package or scope
* Scoped to read or read/write permissions
* Has an expiry date

As mentioned above longer term we are looking at OIDC to remove the needs for tokens and a streamlined GitHub experience to make setting up everything a breeze. All of this is intended to be build on open technologies so that other CI/CD environments can benefit from similar implementation and that we can explore scaling these approaches to other package registries.

None of the above is part of the signing work outlined in this RFC though.

What I can say is that the remainder of this year has a big focus on "package security" for publishing in the same way we've spent the last 9 months extremely focused on "account security"

### Threat model
Our threat model focuses on package hijacking attacks: where a developer with good intentions, writing secure code (i.e. no accidental vulnerabilities) publishes this package to the npm registry.

> The attacker might then gain access to the maintainers login credentials and upload a modified version of the package to the registry causing end-developers who depend on this package to consume malicious code. This maliciously crafted package is often not reflected in the package's source code repository and uploaded directly from the attacker's machine.

Reviewer 3: ```suggestion
The attacker might then gain access to the maintainers login credentials and upload a modified version of the package to the registry under a different, in-range version number causing end-developers who depend on this package to consume malicious code. This maliciously crafted package is often not reflected in the package's source code repository and uploaded directly from the attacker's machine.
```

As such, what we care about in this threat model is whether this provenance information can be forged and whether verification can be bypassed.

The following table highlights the impact of different compromises:

| Compromised infrastructure | Forge provenance                 | Bypass verification |
|----------------------------|----------------------------------|--------------------|
| npm registry               | No                               | Yes                |
| CI/CD                      | Yes                              | No                 |
| Sigstore                   | Yes                              | No                 |
| Network (MITM)             | No                               | Yes                |
| User account (npm)         | No                               | Yes                |
| User account (GH)          | No                               | No                 |
| Trusted builder            | Likely                           | No                 |

> An attacker could forge what goes in the provenance information if the CI/CD provider or Sigstore was compromised. Similarly, if the trusted builder was compromised in some way it's likely the attacker could forge the provenance information during the build. Mitigating these attacks is out of scope in this proposal.

Author: Good point, have not considered this attack vector but we probably want to at least think about it!

> An attacker could bypass verification if the npm registry, npm user or the network was compromised by not providing provenance information or removing it in transit.

Author: > Does this imply that `npm install` would fail open if no provenance info is provided?

Yes I think we need to go with fail open by default until some large portion of packages have provenance.

> Could an option be added to require it?

Yes! I'd love to provide ways for consumers to require provenance and specify policies around which packages or org scopes for example are required 👍  Outlined a similar idea under "Sigstore integration in the npm CLI > Verifying"

> An attacker could bypass verification if the npm registry, npm user or the network was compromised by not providing provenance information or removing it in transit.

Reviewer: It's probably worth reviewing/discussing this in the context of https://github.com/npm/rfcs/pull/182 which has been pending the now-shipped `npm query` that also has plans for policies that are related to audit :)

> An attacker could bypass verification if the npm registry, npm user or the network was compromised by not providing provenance information or removing it in transit.

Reviewer 15: requiring provenance should be sticky, if a package stops providing provenence data it should fail the install and require intervention.

There should be similar warnings for publishers.

> An attacker could bypass verification if the npm registry, npm user or the network was compromised by not providing provenance information or removing it in transit.

Author: > requiring provenance should be sticky, if a package stops providing provenence data it should fail the install and require intervention.

Yes I would love to enforce this somehow. There probably needs to be an escape hatch if you need to move publishing to some platform that isn't supported for example, but should require some hoops to disable, e.g. 2FA must be used to allow publishing a new version without provenance information.

> An attacker could bypass verification if the npm registry, npm user or the network was compromised by not providing provenance information or removing it in transit.

Reviewer 11: note: This is the "SLSA policy" question, which is a bit complicated. See my other comment below (line 365). It might be best to leave this to a follow-up RFC.

> The long-term solution to bypassing verification is enforcing or requiring that all packages have this information set during install time in the npm CLI. It might take years to get to the point where a large portion of npm packages have this information set so we'll need to find ways to get there gradually.

Reviewer 2: What is "this information"?

> We could make it harder to remove provenance information once it's been set by a maintainer. For example, preventing publishes that don't include it if the current latest version does include it. Requiring user 2FA to remove it once it's set.

Reviewer 2: Should "publishes" be "publications"?

Distributing author-managed public keys (PKI) remains a challenge as well as guaranteeing secure handling of private key material. Leaked credentials are a widespread problem among existing PKI implementations, as evidenced by the large number of SSH keys that both [independent researchers](https://www.ndss-symposium.org/wp-content/uploads/2019/02/ndss2019_04B-3_Meli_paper.pdf) and GitHub’s internal secret scanning program have found over the years.

#### Long-lived X.509 certificates
Long-lived X.509 certificates are in widespread use to secure web pages, but they are still cumbersome to manage.

Long-lived certificates have the advantage of being able to chain up to a certificate authority that has intrinsic trust, so beyond authenticating that the artifact signed by the certificate is valid, a consumer can validate that the artifact is from an author that they trust.

> Like other author-managed keys, the developer must keep and secure the private certificate throughout the lifetime of the certificate. Long-lived certificates may also encounter issues with needing to manage revocation in the event of a private certificate leak. While a centralized certificate authority could issue long-lived X.509 certificates to developers as Let's Encrypt has done for web pages, the need to manage revocation across software artifacts in the registry is a problem.

Reviewer 2: s/private certificate/private key/

Nuget, for example, supports [CA-issued code signing certificates](https://docs.microsoft.com/en-us/nuget/create-packages/sign-a-package). Its seen low adoption among the open source community due to several reported [problems](https://haacked.com/archive/2019/04/03/nuget-package-signing/).

#### Short-lived X.509 certificates
Short-lived X.509 certificates eliminate the need to manage keys across a long lifetime and eliminate the need to handle revocations.

Combined with a trusted certificate authority that can create short-lived certificates and log the signatures that took place on them, we can validate any artifacts that were signed by that certificate, and protect against repudiation by maintaining an immutable log of the signing activities.

> 

Reviewer 10: It may be worth mentioning that signed timestamps also support enforcing that a certificate was valid when the signature was created.

There is some good documentation around this by Authenticode - https://docs.microsoft.com/en-us/windows/win32/seccrypto/time-stamping-authenticode-signatures

Author: Thanks included this 👍

Today, this means adding direct vendor support for commercial CI/CD vendors. We realize that this isn't ideal as it centralizes support to a handful of vendors. We believe this compromise is the best option we have given our goals around building and publishing in the open, respecting maintainer pseudonymity and avoiding maintainer-managed keys or secrets.

### Authorizing a release
A provenance attestation captures information about the source code repo, revision and build information. This information on its own does not capture if the release was accepted by the registry as coming from an authorized user.

While inclusion in the registry is proof that the package was published by an authorized actor, no verifiable attestation exists that captures the details of the decision taken by the registry to allow the publish step.

As part of authorizing publishing of the package, the registry should sign a statement (release attestation) about accepting the package release and publish it to Rekor (public ledger).

> Website CAs currently use a similar approach when they publish all issued certificates to a [transparency log](https://certificate.transparency.dev/).

Reviewer 10: nit: WebPKI instead of "website"

### CI/CD OIDC provider support

> Today only GitHub Actions is fully supported by Fulcio. We’d like to see support added for any public CI/CD service that can meet these requirements:

Reviewer: If this means that GitHub Actions are the only way to publish verified packages, I'm a hard and vocal -1 on this until that's not the case. There are extremely valid reasons to not use GitHub Actions that haven't been addressed in the years it's been out, and there are major packages in the ecosystem that do not use GItHub Actions for those reasons.

Reviewer: (from the following context that does not seem to be the case for npm to ship this, but I am going to leave this here to ensure that perspective is logged)

Author: 100%, GitHub Actions is the only provider that fully supports the requirements from Fulcio today but I'd like to see other CI/CD providers like GitLab and Circle be supported before this functionality launches for npm.

Reviewer 15: Would it be feasable to allow any OICD provider to include signatures with published npm packages and let the npm cli configure which providers it trusts? npmjs.com can choose to only display verified badges for Github Actions and other trusted providers but other communities could choose to trust providers of their choice.

Author: > Would it be feasable to allow any OICD provider to include signatures with published npm packages and let the npm cli configure which providers it trusts? 

The current proposal relies on the OIDC Identity Providers trusted by Fulcio, which is currently defined here: https://github.com/sigstore/fulcio/tree/main/federation

On top if this I could also see npm providing some way for you to write your own policies of which of these providers to trust. The hope is that the list of supported and trusted providers in Fulcio grows significantly over time as its vendor neutral and open source.

One idea to make it easier to support more providers in Fulcio would be to come up with a spec for ID token claims, so any provider that supports some officially supported set of claims could be supported more easily in Fulcio.

- OIDC ID tokens identifying the current workflow/run/build.
- ID token with a custom audience (needs to be set to `sigstore` for Fulcio to accept it).
- Claims about the code repository, commit, build and actor (see a detailed table of required ID token claims below)

As of today, Google Cloud Build supports custom claims but neither GitLab Pipelines nor CircleCI offers this. [GitLab](https://gitlab.com/gitlab-org/gitlab/-/issues/356986) is planning to add support for a custom audience.

> The plan is to start with direct vendor support in Fulcio. Fulcio will need to be patched to produce certs containing custom claims data from Circle CI, GitLab and Google Cloud Build.

Reviewer 10: nit: certificates, not "certs"

> We will also work with OpenSSF and other package ecosystems adopting [Sigstore](https://www.sigstore.dev/) on an official claims set that any provider can implement. Once we have this each vendor could be added to Fulcio by simply specifying the standard OIDC client configuration.

Reviewer 10: (From a Fulcio maintainer) This would be great!

| GitLab Pipelines   | yes          | no              | Subject is the project path. See [documentation](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html).                                                                                                                   |
| CircleCI           | yes          | no              | Uses [workload ID](https://circleci.com/docs/2.0/openid-connect-tokens/) that includes a notion of “user ID” in the Fsubject claim. There is a [feature request to make this configurable](https://circleci.canny.io/cloud-feature-requests/p/customizable-audience-claim-in-oidc-tokens) |
| Google Cloud Build | yes          | yes             |                                                                                                                                                     |
| Bitbucket          | yes          | no              | Is it used for OSS?                                                                                                                                                    |

> | TravisCI           | no           | -               | Is it used for OSS?                                                                                                                                                    |

Reviewer 10: Jenkins comparison?

| Actor id        | The ID of the account that initiated the build.                           | actor_id         | user_id                                 | See above comment on “Actor”                                   |
| Build name      | The name of the build.                                                    | workflow         | pipeline_id (i.e no name, only numeric) | N/A                                                            |
| Build id        | The ID of the build.                                                      | run_id           |                                         |                                                                |
| Build reference | Ref to an external workflow if one is being used, e.g. a trusted builder. | job_workflow_ref | N/A                                     | N/A                                                            |

## Implementation details
### Sigstore integration in the npm CLI
Generating build provenance during publish will be opt-in using a command line flag and only run in supported CI/CD systems.

> When enabled on publish, the npm CLI will interact with Sigstore through a reverse proxy on npmjs.org: certifying the authenticity of the CD/CD workflow identity token ([fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/)) and verifying that signing happened during the time the id certificate was valid ([rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/)).

Reviewer: > npmjs.org

IIRC npmjs.com is used as the primary source now?

The signature will be placed on the public Rekor ledger for transparency but a detached copy will be uploaded to the npm registry and used for verification. The detached copy includes proof that it was verified and included by Rekor so we don't need to query it again during verification.

Signature verification will happen transparently during `npm install` and `npm audit signatures` so needs to meet strict performance and reliability uptime requirements as it will be performed for every single npm install. The provenance attestations/signatures will be fetched in bulk from the registry, similar to how advisories are fetched in bulk for all dependencies in the current install. This means verification will not rely on any Sigstore infrastructure to work.

> The signing and verification logic will be implemented in a [sigstore-js](https://github.com/sigstore/sigstore-js) client library using Node’s crypto library. Similar clients exist in [Go](https://github.com/sigstore/cosign), [Python](https://github.com/sigstore/sigstore-python) and [Java](http://github.com/sigstore/sigstore-java) and we plan to follow their lead on implementation details.

Reviewer: ```suggestion
The signing and verification logic will be implemented in a [sigstore-js](https://github.com/sigstore/sigstore-js) client library using Node.js's crypto library. Similar clients exist in [Go](https://github.com/sigstore/cosign), [Python](https://github.com/sigstore/sigstore-python) and [Java](http://github.com/sigstore/sigstore-java) and we plan to follow their lead on implementation details.
```

The npm CLI will rely on the Sigstore JS client library to generate the provenance attestation on publish and to verify the detached signature bundle that's downloaded from the registry.

The Sigstore client library will include direct vendor support for requesting an OIDC id token from supported CI/CD systems, similar to the existing client libraries. It will also include all the required functionality to interact with Fulcio and Rekor, as well as verify the [Sigstore root keys](https://github.com/sigstore/root-signing) using a [TUF](https://theupdateframework.github.io/) client.

#### Fetching Sigstore hosts from npmjs.org
The npm CLI will fetch Sigstore config from npmjs.org that contains the public Fulcio and Rekor hosts by default:
```
GET https://registry.npmjs.org/-/npm/v1/sigstore-config

{
  "sigstoreConfig": {
    "fulcioUrl": "https://fulcio.sigstore.dev",
    "rekorUrl": "https://rekor.sigstore.dev"

>   }

Reviewer 10: Also configure the GCS bucket location of the Sigstore TUF root

> The npm CLI will fetch Sigstore config from npmjs.org that contains the public Fulcio and Rekor hosts by default:

Reviewer: Is this something that will be configurable? Alternatively, if I intentionally don't allow the npm CLI to access npmjs.org, can my registry provide this?

Author: Yes the idea is that the CLI attempts to fetch this config from the configured registry, so does not need to be npmjs.org. Will clarify 👍

}
```

This would allow future extensions where maintainers can configure the instance being used, making sure the release attestation gets placed on the same Rekor instance as the build attestation. The release attestation will be generated by the registry when the package has been successfully authorized and published.

#### Publishing

> Maintainers will need to enable build provenance generation with a command line flag: `npm publish --build-provenance` in addition to running it in a supported CI/CD system. The command will error with an error message if this is not run in a supported build system.

Reviewer: would `npm publish --with=build-provenance` be more in-line with how the DX of the npm CLI is currently evolving?

Reviewer: it might also be preferable to directly call out the name in case the system changes in the future: `npm publish --with=sigstore-build-provenance`

Once enabled, the CLI will perform the required steps to generate the provenance attestation and upload this to the public Rekor ledger for transparency and a detached copy that contains proof of inclusion from Rekor will be stored in the npm registry.

##### What goes on the public Rekor ledger?
Only public npm packages will be signed and published to the public [rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview) ledger by default.

> Privately scoped packages, or packages from private repositories, will not be signed or published to the public ledger. This will be determined by interrogating the npm registry at publish time. Users will be able to override this similar to the `cosign --force` command by passing an argument to `npm publish`, e.g. `--force-build-provenance`.

Reviewer 2: Is there a potential dependency confusion attack here? Ie. can I get a private package to publish to rekor if I upload a package of that name to the public registry?

Author: Oh yeah you mean a case where you have some internal corp dep pulled from a private registry and then someone finds this name and creates a public version? Not sure we can do much to mitigate the creation of these. Do you see an added risk being able to create these?

An issue could be getting the wrong provenance attestations during verification if we just used the package name and version. We can mitigate this by querying for the attestations using a shamsum of the installed package's tarball.

Reviewer: I think what @jchestershopify means is if I have package `x` internally and someone publishes `x`, the RFC says that npm will interrogate the registry for `x` and get the internal package signed accidentally because `x` exists on the public registry.

Author: When you say internal package `x` you mean on some other registry right? I think in this case the signature would be for the public package `x` and associated via the shasum of the tarball, which won't match the internal package `x`. Maybe I'm missing something though?

Author: We're also planning on creating a "release attestation" on the npm registry that gets uploaded to the public Rekor ledger. This is a signed statement that the npm registry accepted and authorized the publish for public package `x` on top of the "build attestation" that gets created during publish in the CI/CD system, which links the source and build to the package.

> Privately scoped packages, or packages from private repositories, will not be signed or published to the public ledger. This will be determined by interrogating the npm registry at publish time. Users will be able to override this similar to the `cosign --force` command by passing an argument to `npm publish`, e.g. `--force-build-provenance`.

Reviewer: What happens when the npm CLI is intentionally blocked from reaching out to the npm registry?

Author: The idea is that any npm registry could implement the required changes to support this signing flow. The CLI would try and get the sigstore config from the regitry you are publishing to, not just the official registry.

> The entry will be stored and retrieved using the shasum of the package tarball. This will make sure we don't find any signatures for a  deleted package that has been resurrected with the same name and version.

Reviewer: ```suggestion
The entry will be stored and retrieved using the shasum of the package tarball. This will make sure we don't find any signatures for a deleted package that has been resurrected with the same name and version.
```

There will be no way of deleting an entry if this was published unintentionally to the ledger as it's immutable by design to prevent tampering. The practical consequence of publishing to the ledger is that information about your package (name, version), repository (commit, branch, owner) and workflow (name, id) are uploaded to the public ledger. A possible workaround would be to rename the leaked repo, make it private and delete the old one.

##### Detailed steps to publish
Interaction between the npm CLI tool running in CI, the npm registry, and the Sigstore components Fulcio and Rekor:

> ```mermaid

Reviewer 10: Do you want to include the OIDC request in this diagram?

Author: Good spot, updated

    participant fulcio
    participant rekor
    participant npm_registry
    npm_cli->>fulcio: Request signing certificate
    fulcio->>npm_cli: Signing certificate
    npm_cli->>rekor: Upload signed build provenance attestation
    rekor->>npm_cli: Rekor entry
    npm_cli->>npm_registry: Publish package and upload Rekor entry
    npm_registry->>fulcio: Request signing certificate
    fulcio->>npm_registry: Signing certificate
    npm_registry->>rekor: Upload signed release attestations
    rekor->>npm_registry: Rekor entry
    npm_registry->>npm_registry: Attach rekor entry to published package
    npm_registry->>npm_cli: Success
```

1. `npm publish --build-provenance` detects execution within a supported CI/CD environment and initializes token exchange to get a JWT id token from the CI/CD system's OIDC identity provider.
2. npm CLI creates a new ephemeral key pair, requests a certificate from Fulcio (CA) and authenticates via the JWT id token.
3. A SLSA build provenance attestation is created based on information available within the JWT and CI system and signs it with the ephemeral key.
4. npm CLI uploads the completed attestation record on Rekor for public packages.

> 5. npm registry creates a release attestation, stating that the version has been accepted and published, and places the record on Rekor for public packages.

Reviewer 10: (Maybe I missed this) Can you motivate why we want the registry to upload a release attestation in addition to the artifact attestation? What does this second attestation add?

Author: Yes the way I see it is that the artifact or build attestation links the package an authenticated build identity. The release attestation will be created by the implementing registry, stating that the user who is publishing this package is in fact authorized to do so and the package has been published.

This links the build identity to an authorized publish. In future if we supported staged releases for example, the release attestation would only be created when the staged release was approved. It also creates an audit trail for public npm publishes, potentially allowing prudent organisations to monitor the log for any publishes to packages or org scopes they own.

Reviewer 10: Would you consider publishing all events that occur for a package to the log? For example, in addition to publishing the release, what about yanking? What about updating critical metadata (such as who the maintainer is)?

Reviewer 2: When we looked at this question for RubyGems (before we started working with other ecosystems through the OpenSSF's Securing Software Repos WG), we identified these operations as candidates for recording to the transparency log:

* Registration
* Sign during build
* Push
* Publish
* Yank
* Owner added
* Owner removed
* Owner changed email address
* Owner enabled MFA
* Owner disabled MFA
* Endorsement by non-owner
* Ownership call
* Ownership request
* Ownership transfer
* Deletion
* Resurrection of name

Reviewer 2: More generally I see publishing a rich set of events to the log as creating a safe independent history of sensitive events. This provides the possibility of detection when the repo DB or blob bucket are compromised, because it would be out of sync with the transparency log.

Author: > Would you consider publishing all events that occur for a package to the log?

Yes would like to explore this more. The obvious one that jumps out is publishing an entry for deletions/yanks and malware takedowns.

It's probably a good idea to go slow adding these to make sure we have a good story for knowing what to do with it once we have it, e.g. during verification, and making sure we don't start spamming Rekor with too many events.

Reviewer 2: @feelepxyz My ask is that we coordinate a bit on the schema for such events through the Securing Software Repos WG. One of my north stars is for my colleagues in AppSec to only need to install a single log monitor for all the ecosystems we rely on.

Author: > @feelepxyz My ask is that we coordinate a bit on the schema for such events through the Securing Software Repos WG.

Yes sounds good to me 👍

Author: @jchestershopify I added a section at the bottom about publishing additional events and aligning with the securing repos wg 👍

#### Verifying
The package integrity captured by the provenance attestation will be verified by the npm CLI. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

This will be staged rollout over several npm CLI releases as changing the default behavior of `npm install` is a breaking change:

> 1. The existing `npm audit signatures` command will start verifying build signatures, possibly adding an opt-in flag to audit these signatures during `npm install`, e.g. `--audit-build-provenance`

Reviewer: rather than forever adding standalone flags, it'd be nice to think about the long-term DX of flags. Adding `npm install --audit=build-provenance` rather than a single flag would be preferable IMO.

Author: > `npm install --audit=build-provenance`

Nice! This looks preferable to me to 👍

> 2. Start auditing build signatures by default when running `npm install`. This will only check packages that have build attestations and ignore the rest. Allow opting out of this behavior with `--no-audit-build-provenance` or `--no-audit`.

Reviewer: same comment as above but `--no-audit=build-provenance`

> 3. Allow users to opt-in to a strict mode to require packages to be published with build provenance information, e.g. `npm install --require-build-provenance`. Initially, this might only be usable by maintainers that only depend on their dependencies. It will take a long time for this to be feasible for the average install as all transitive dependencies would need to include build provenance on the installed version. We could allow policies that make gradual enforcement possible, e.g. only require build provenance on packages published after some minimum date.

Reviewer: same as above :P

> Signature verification needs to be performant and have a negligible impact on `npm install`. We believe the main bottleneck here will be the network, downloading the signatures to verify. The crypto operations involved in verification are negligible on modern hardware. We propose the CLI makes a bulk API request to the npm registry, similar to how advisories are fetched, for all provenance attestations for a given install.

Reviewer: Committing to a benchmark for this to be able to spot regressions would be great as well 👀

Author: Yes I like this idea 👍

##### Detailed steps to verify
1. npm CLI downloads the provenance and release attestations for each installed package (the download will be done in bulk for all installed packages)
    1. Verify the build provenance attestation is signed by a certificate created for the builder/actions run.
    2. One Rekor entry proves that the build provenance attestation was placed on the public ledger.
    3. One Rekor entry proves that the npm package was accepted and released by the npm registry (the Rekor entry contains the package’s url (PURL), sha512 digest and is signed with a certificate for the npm registry).
    4. Verify both of the certificates are issued by Fulcio and identifying different workloads (builder and the npm registry).

> 2. npm CLI verifies that the Rekor entries, certificate and certificate chain in the build provenance are trusted. The [sigstore-js](https://github.com/sigstore/sigstore-js) client includes a TUF client that's used to verify, and optionally update, the public keys for both Rekor and Fulcio.

Reviewer 10: Can you also verify that the Fulcio certificate was issued to a certificate transparency log? Like WebPKI, Fulcio embeds a Signed Certificate Timestamp (SCT) in the certificate as a promise of inclusion in a log.

There is a separate key in the TUF targets for this.

Author: Yes I believe we're working on verifying the SCT in https://github.com/sigstore/sigstore-js (there's a bunch of verification steps missing in there atm)

9. npm CLI verifies that the certificate used to sign the package digest matches the package name and repository in the build provenance’s config source section.

There are two obvious failure scenarios we need to handle when verifying:
- The network call to fetch attestations fails: This should probably error and instruct users to re-run `npm audit signatures` or `npm install`.
- Something genuinely unexpected happened and the attestation signature doesn't match the package that's been downloaded: This should fail the audit or install instruct users to manually inspect the failing package(s) to determine if they might have been tampered with.

### Supporting third-party npm clients
Third-party npm clients, e.g. `yarn` and `pnpm`, should be able to implement build provenance generation using the [sigstore-js](https://github.com/sigstore/sigstore-js) client library. This library will contain all the functionality required to extract build provenance information, generate the OIDC id token on supported CI/CD systems, authenticate it against Fulcio, generate the signature and upload it to the public Rekor ledger.

> We're planning on providing detailed documentation on how to wire up the Sigstore client library in any npm client.

Reviewer: ```suggestion
We're planning on providing detailed documentation on how to wire up the Sigstore client library in any JavaScript package manager.
```

Reviewer: not necessarily this, but this should be altered as the language is confusing.

At a high level, third-party registries will need to support the following:
- Generate release attestations
- Accept build provenance attestation documents on publish, alongside the packument (`package.json`) and tarball
- Storing the attestations in the registry
- Providing a query endpoint for retrieving attestations

We're planning on providing detailed documentation on the required changes.

### Non-falsifiable provenance using a trusted builder
To ensure the provenance information hasn't been tampered with during the build, e.g. by a malicious dependency, we need to have a way to run a build and publish in a trusted build environment that can't be directly accessed or modified by the source repository where it's being executed.

Non-falsifiable provenance information is a [requirement to reach SLSA level 3](https://slsa.dev/spec/v0.1/requirements#non-falsifiable) (framework for securing the supply chain).

Using a trusted builder to publish an npm release will ensure a higher level of build integrity. Using it will be optional, but highly recommended. Verification of build provenance will include checks to determine if a trusted builder was used.

> At a high level, the trusted builder will checkout the git repo, install dependencies and execute `npm run build` followed by `npm publish`. Maintainers can only control what is run during `npm run build` not add or remove steps from the trusted builder. Provenance generation will be run in a separate workflow step that's isolated from the build and install.

Reviewer: `npm run build` here should be configurable. This is not the build command that everyone uses, and could potentially add significant overhead for projects with existing workflows.

Reviewer 15: Only run `npm publish` and allow extension via `prepublishOnly`.

Reviewer 3: … or prepare, or prepack, or any other relevant lifecycle event, presumably?

The [slsa-github-generator](https://github.com/slsa-framework/slsa-github-generator) is an example of a trusted builder using GitHub Actions [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows).

### Provenance attestations
The process of creating a provenance attestation is dependent on the current CI/CD system being used as there's currently no standardized way of retrieving this information. Different methods and variables will need to be used to gather information about the current branch, commit version, job id, actor etc.

Existing tools like [witness](https://github.com/testifysec/witness) and Sigstore's [cosign](https://github.com/sigstore/cosign) CLI implement direct support for various CI/CD systems, so the [sigstore-js](https://github.com/sigstore/sigstore-js) client used in the npm CLI should follow that pattern and be able to figure out in which environment it’s running in.

> This means that it's not required to run the trusted builder to include build provenance on publish. This would also allow users to use private instances of Fulcio and Rekor.

Reviewer 10: It may be worthwhile to link to https://github.com/slsa-framework/slsa-verifier in the RFC for a way to verify provenance.

Author: Nice, provided a link to this under "Prior Art"

> Existing tools like [witness](https://github.com/testifysec/witness) and Sigstore's [cosign](https://github.com/sigstore/cosign) CLI implement direct support for various CI/CD systems, so the [sigstore-js](https://github.com/sigstore/sigstore-js) client used in the npm CLI should follow that pattern and be able to figure out in which environment it’s running in.

Reviewer: Should this be a hard blocker for landing sigstore-js?

Reviewer: Sorry, to be more clear: should the direct support for various CI/CD systems that matches what witness and cosign do be a hard requirement prior to landing it in the npm CLI?

Author: The bar for this is low atm, only GitHub Actions is supported by Fulcio at the moment so the CLI should match this and keep up with any new CI/CD providers that are supported by Fulcio as they get added to https://github.com/sigstore/fulcio/tree/main/federation

Author: Clarified this 👍

| builder.id                         | https://github.com/npm/.github/workflows/slsa@refs/head/main/<git_sha>                                                                                                                                             |
| buildType                          | https://npmjs.org/slsa/generator@v1                                                                                                                                                                                |
| invocation.configSource.uri        | https://github.com/onwer/repo/refs/head/<branch>                                                                                                                                                               |
| invocation.configSource.digest     | Git SHA                                                                                                                                                                                                            |

> | invocation.configSource.entryPoint | workflow name                                                                                                                                                                                                    |

Author: Thanks for your context on generating the provenance statement! Very helpful.

In GHA this information is available via environment variables as well as the OIDC id token (JWT). These variables should not be included in the build invocation environment, as they don't affect the build.

#### Environment variables needed during provenance generation
The environment variables below are needed to create the materials section, `configSource` and the build invocation id in the provenance attestation. Some variables listed below are also available as claims in the OIDC id token (JWT), reading those values from the JWT is preferable as it's authenticated.

| What              | GitHub                     |
|-------------------|----------------------------|
| Run id            | GITHUB_CONTEXT.run_id      |
| Run attempt       | GITHUB_CONTEXT.run_attempt |
| Workflow name     | GITHUB_CONTEXT.workflow    |
| Server URL        | GITHUB_CONTEXT.server_url  |
| Repository        | GITHUB_CONTEXT.repository  |
| Ref               | GITHUB_CONTEXT.ref         |
| Git SHA           | GITHUB_CONTEXT.sha         |
| OS image          | ImageOS                    |
| OS image version  | ImageVersion               |

> | Host architecture | RUNNER_ARCH                |

Reviewer: This seems GitHub-centric. Other platforms should be added with `???` for unknown values.

#### Signature envelope
The signed build provenance attestation must be stored and transferred together with its signature. The envelope carries this functionality. We plan to support [DSSE envelopes](https://github.com/secure-systems-lab/dsse), and preferably [COSE envelopes](https://datatracker.ietf.org/doc/html/rfc8152) as well.

For a more detailed look into different envelope schemas, see this [comparison of signing envelope formats](https://docs.google.com/document/d/18YVGA4mq45wfUkWrAqWkymzdHRcXxlwINKXnEp86L0w/edit#heading=h.3v6vw85pjg5e).

### Release attestations
When a package is successfully published, the npm registry will "attest" the release and upload this to Rekor. This includes a signature that links the repository to the package name. This must be verified by the CLI when during verification.

This can be achieved by publishing the package signature to Rekor by using the proposed [“package entry”](https://github.com/sigstore/rekor/pull/805) log type.

The release attestation will contain the following claims, linking the build to the release:
- Package org/name
- Package version
- Registry signature
- Tarball shasum

## Prior Art
### Sigstore cosign CLI

> The [cosign](https://github.com/sigstore/cosign) CLI tool supports signing arbitrary blogs and could be used to sign the package tarball manually but requires significant setup from maintainers and offers no simple way for consumers to verify signatures unless you know you to query for the package signature on Rekor.

Reviewer 2: s/blogs/blobs/

- **OpenID Connect ([OIDC](https://openid.net/specs/openid-connect-core-1_0.html)) identity or ID token**: Verifiable information that the user or identity has been authenticated by a OIDC provider. The ID token is a JSON Web Token (JWT).
- **OpenID Connect ([OIDC](https://openid.net/specs/openid-connect-core-1_0.html)) identity provider**:  Authenticates users and issues verifiable id tokens that includes identity information, e.g. email. Examples include Google or GitHub.
- **Package hijacking attack**: Where a malicious version of an existing open source package gets uploaded to the registry. The [eslint-scope](https://eslint.org/blog/2018/07/postmortem-for-malicious-package-publishes) attack was a notable example of this kind of attack, and it frequently occurs due to compromised npm credentials but can also happen due to compromised builds or CI/CD.
- **[Rekor](https://docs.sigstore.dev/rekor/overview)**: Public immutable tamper-resistant ledger of signed software artifacts. Verifies that the Fulcio certificate was valid at the time of signing.

> - **Release attestation**: Verifiable metadata stating that a new package release has been published on the npm registry

Reviewer 2: Nit: missing period.

    "tufRepository": "https://sigstore-tuf-root.storage.googleapis.com",
    "tufRoot": "root.json"
  }
}
```

This would allow future extensions where maintainers can configure the instance being used, making sure the release attestation gets placed on the same Rekor instance as the build attestation. The release attestation will be generated by the registry when the package has been successfully authorized and published.

#### Publishing
Maintainers will need to enable build provenance generation with a command line flag: `npm publish --with=build-signatures` in addition to running it in a supported CI/CD system. The command will error with an error message if this is not run in a supported build system.

Once enabled, the CLI will perform the required steps to generate the provenance attestation and upload this to the public Rekor ledger for transparency and a detached copy that contains proof of inclusion from Rekor will be stored in the npm registry.

##### What goes on the public Rekor ledger?
Only public npm packages will be signed and published to the public [rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview) ledger by default.

> Privately scoped packages, or packages from private repositories, will not be signed or published to the public ledger. This will be determined by interrogating the npm registry at publish time. Users will be able to override this similar to the `cosign --force` command by passing an argument to `npm publish`, e.g. `--force-build-signatures`.

Reviewer 15: Why is `--force-build-signatures` required when `--with=build-signatures` is not enabled by default?

Author: > Why is `--force-build-signatures` required when `--with=build-signatures` is not enabled by default?

This was meant for private packages, that by default don't upload provenance information on the immutable public ledger Rekor, even if you say `--with=build-signatures`. Not sure we need this unless we have users who actually want to upload this information about private packages.

Reviewer 8: Would it be possible to use something a bit more descriptive, like `--force-publish-signatures` instead? It would eliminate confusion.

We could make it harder to remove provenance information once it's been set by a maintainer. For example, blocking a npm publish that doesn't include it if the current latest version does include it. Requiring user 2FA to remove it once it's set.

Maintainers and developers consuming packages could also opt-in to requiring all their dependencies to include provenance information once their dependencies include this.

## Rationale and Alternatives
While no existing solution addresses all of the goals outlined above at this time, the [Sigstore](https://www.sigstore.dev/) project and its use of short-lived X.509 certificates is closest to our stated goals for the vast majority of developers since it reduces the complexity of key management, and removes the need for revocation scenarios.

Using OIDC to handle authentication of workloads requesting a short-lived X.509 certificate allows this solution to be vendor neutral and work consistently across many source repositories and CI systems.

> ### Signing methods

Reviewer 11: Another alternatives is Long-lived CI/CD managed keys. For example, I believe Google Cloud Build signs all of its provenance attestations with some fixed public key. I believe this would model still address most or all of the requirements. Under the current proposal, NPM already has to maintain a list of trusted builders; using a fixed key per builder would just add one more piece of data to that list. Maintenance might be a bit more difficult, but I think it's worth adding to this list.

Author: > Another alternatives is Long-lived CI/CD managed keys

Yes! We actually floated this idea early on but ran into questions around key revocation and distribution. Will add this as a alternative 👍 

I get the feeling GCP is minting one key-pair per project or repository from [glancing the docs](https://cloud.google.com/build/docs/securing-builds/view-build-provenance) but might be wrong?

Reviewer 11: Oops, I think you're right. Let's just remove GCB as an example and speak hypothetically.

| Actor           | The account that initiated the build.                                     | actor            | user_login or user_email                | Need to parse subject claim.                                   |
| Actor id        | The ID of the account that initiated the build.                           | actor_id         | user_id                                 | See above comment on “Actor”                                   |
| Build name      | The name of the build.                                                    | workflow         | pipeline_id (i.e no name, only numeric) | N/A                                                            |
| Build id        | The ID of the build.                                                      | run_id + run_attempt           |                                         |                                                                |
| Build reference | Ref to an external workflow if one is being used, e.g. a trusted builder. | job_workflow_ref | N/A                                     | N/A                                                            |

> ## Implementation details

Reviewer 11: I find this section misleading because there are actually **two** different modes to publish, but that fact is not obvious (at least to me):

1. The "sigstore integration in the npm CLI", i.e. the CLI does the signing by grabbing an OIDC token from the CI/CD system.
2. The "Non-falsifiable provenance using a trusted builder", i.e. the CI/CD does the signing with no integration with the CLI.

I suggest calling this out up front and perhaps having a diagram illustrating the two different models. The former has weaker security guarantees around the accuracy of the provenance.

Author: @MarkLodato yes this is a great callout. I'll clarify this up front 👍

    rekor->>npm_cli: Rekor entry
    npm_cli->>npm_registry: Publish package and upload Rekor entry
    npm_registry->>fulcio: Request signing certificate
    fulcio->>npm_registry: Signing certificate
    npm_registry->>rekor: Upload signed release attestations
    rekor->>npm_registry: Rekor entry
    npm_registry->>npm_registry: Attach rekor entry to published package
    npm_registry->>npm_cli: Success
```

1. `npm publish --with=build-signatures` detects execution within a supported CI/CD environment and initializes token exchange to get a JWT id token from the CI/CD system's OIDC identity provider.
2. npm CLI creates a new ephemeral key pair, requests a certificate from Fulcio (CA) and authenticates via the JWT id token.
3. A SLSA build provenance attestation is created based on information available within the JWT and CI system and signs it with the ephemeral key.
4. npm CLI uploads the completed attestation record on Rekor for public packages.
5. npm registry creates a release attestation, stating that the version has been accepted and published, and places the record on Rekor for public packages.

#### Verifying

> The package integrity captured by the provenance attestation will be verified by the npm CLI. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.

Reviewer 11: It's not clear to me what security value this provides, absent a policy that requires builds to come from a specific source repo. IIUC, this says: "if there is provenance, it has to come from one of the trusted builders, but it can be built from any source." I could see value if that provenance is then displayed to the user or otherwise acted upon, but IIUC it is not. I don't think this will stop any attacks because the attacker can just built "verifiably" from a fork, then delete the fork.

Could you explain the purpose of this check?

(To be clear, I see the value in verifying on the NPM website, and in laying the groundwork for future policy checks. I was just wondering what the point is in the check as written here.)

Author: This is a good point. Just verifying these signatures doesn't say if the provenance has changed but it does verify the package hasn't been tampered with since build. What we can do is start enforcing the presence of provenance information (making it sticky) once it's been set on a package and possibly prevent a publish if this changes, unless you unlink it from your npm dashboard using 2FA for example.

It would also be great to figure out how we could bubble up changes to consumers somehow through the CLI when running these commands. One idea would be to warn on any changes when doing package upgrades in the CLI as we could keep track of before and after. I also envision Dependabot PRs highlighting these changes.

Reviewer 11: Yeah, displaying a warning on provenance change during upgrades would be a great next step. Maybe add that brief explanation and idea to the RFC?

(A subsequent step could be requiring explicit authorization for provenance changes, such as by recording the source repo in the package.lock file. In my opinion that's a decent idea for policy.)

Reviewer 12: Don't want to get too deep into the weeds, but things on my end we've done in the past is have a separate source attestation, or trusted signature against the source. Then the policy becomes both a verification that there is an attestation against the build and then looking in the SLSA provenance for `materials`.

This makes some assumptions though that the builder can't lie about the source it came from. So we need to make sure the GHA flow doesn't allow for the package maintainer piece to lie about the code.

- Avoids developer-managed keys (as there's no good way to offer trust to the community given the challenges of distributing public keys). This, along with opt-in signing, adds near-zero initial and ongoing overhead for open source maintainers.
- Maintainers can opt-in to linking their packages to the source and build it originated from when publishing using `npm publish` for public packages.
- Incentivizes maintainers to build in the open because of the strong guarantees that this offers.
- Verification happens transparently on `npm install` without the need to obtain or manage additional tools or keys.
- Verification should have a negligible performance impact on `npm install`.
- Verification should be performed without depending on any third-party systems other than the registry.
- Compatible with third-party npm clients, e.g. `yarn` and `pnpm`.
- Should allow third-party npm registries, e.g. GitHub Packages, Artifactory and Verdaccio to follow suit and implement similar interfaces.
- Should be maintained with >99.9% uptime so that developers are not blocked from publishing new packages.

> - Should allow future extensions to support centrally managed signing authorities such as certificates managed by an enterprise and inner source within air-gapped enterprise environments.

Reviewer 8: It it possible to run a private rekor (transparency ledger that drives sigstore) instance and pair it up to SPIFFE/SPIRE to achieve this. Not sure how difficult this process is though. If this is an important use case, we can work on simplifying this path in the upstream projects.

Author: > It it possible to run a private rekor (transparency ledger that drives sigstore) instance and pair it up to SPIFFE/SPIRE to achieve this. Not sure how difficult this process is though. If this is an important use case, we can work on simplifying this path in the upstream projects.

Yes I have a vague understanding that something like this should be possible. It's not an immediate goal to support this use case, I'd like to make sure there is some path to get there in future that won't require major reworking.

- Supports signing identities that don't expose any personally identifiable information (PII) about maintainers, e.g. emails.
- Avoids developer-managed keys (as there's no good way to offer trust to the community given the challenges of distributing public keys).

Note that nothing in this current proposal precludes us from offering additional signing methods in the future as they become available and mature.

## Risks of adopting Sigstore
Sigstore infrastructure and tooling are currently not considered production ready and are run on a best-effort basis. The Sigstore project is working towards a [General Availability](https://blog.sigstore.dev/an-update-on-general-availability-5c5563d4e400) release later this year but we don't yet know what these guarantees will look like yet.

As such, there are several risks of adopting Sigstore for npm that are worth calling out:
- No buy-in from the broader open source npm community.
- Sigstore's public and free-to-use infrastructure is not able to meet our uptime requirements (>99.9%).
- Sigstore is not able to secure the funding required to run the infrastructure with production-grade support, SLOs, and on-call rotation.

> To mitigate some of these risks, GitHub is planning to work directly with Sigstore to define and support robust production-grade SLOs for uptime and reliability that meet the requirements for npm. This will involve being on-call and wearing the pager for production services.

Reviewer 8: Does this scope only cover GitHub's infrastructure in how they interact with Sigstore, or are they committing to help maintain the infrastructure through their involvement with the OpenSSF?

Author: @fkautz this will cover the public Sigstore infrastructure. Not sure exactly what the engagement model is at the moment with OpenSSF/Sigstore. @trevrosen from GitHub is now on the Sigstore technical steering committee so might know more about this?

Reviewer 16: @fkautz sigstore is maintained and operated by a collection of folks who work for several different entities: Google, Chainguard, RedHat and (with this RFC and the run-up to GA) GitHub. Folks from GitHub have been contributing patches and architectural expertise to sigstore for some time now and will be involved in guaranteeing operational excellence for the sigstore public good network going forward.

- The email might not match the one used on the registry making it hard to validate that it's from a legitimate maintainer.

##### CI/CD workload identity provider
For instance, Github Actions (supported by Sigstore today) or any other CI/CD system that can provide OIDC workload identities with required claims.

**Pros**
- Support for pseudonyms/privacy: does not require public identities other than what already corresponds to the package (e.g. repository).
- Links the package to the source repository and includes information about and how it was built: commit SHA, workflow run id etc.
- Vendor-neutral: supports any CI/CD system supporting OIDC workload identities.
- OIDC workflow identities could also be used to authenticate the package publish (restricting publishing to a specific source code repo and removing the need for long-lived access tokens).

**Cons**
- Limited support today: Only GitHub Actions is fully supported in Fulcio today.
- Friction: Requires publishing from supported CI/CD providers.
- CI/CD providers will need to operate an IdP and get it integrated with Sigstore.

> Using workload identities along with short-lived X.509 certificates is currently the only solution that meets our key requirements:

Reviewer 12: FRSCA, an OpenSSF project, supports this through SPIFEE/Spire: https://github.com/buildsec/frsca -- We could work to integrate a root of trust associated with an OIDC as well.

|--------------------|--------------|-----------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| GitHub Actions     | yes          | yes             | Uses “workflow run” as the subject                                                                                                 |
| GitLab Pipelines   | yes          | no              | Subject is the project path. See [documentation](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html).                                                                                                                   |
| CircleCI           | yes          | no              | Uses [workload ID](https://circleci.com/docs/2.0/openid-connect-tokens/) that includes a notion of “user ID” in the Fsubject claim. There is a [feature request to make this configurable](https://circleci.canny.io/cloud-feature-requests/p/customizable-audience-claim-in-oidc-tokens) |

> | Google Cloud Build | yes          | yes             |                                                                                                                                                     |

Reviewer 17: Cirrus CI also supports OIDC tokens and now allows to override the audience (see https://github.com/cirruslabs/cirrus-ci-docs/pull/1058). Will appreciate if Cirrus CI can be also included.  We also support a [pretty similar to GHA set of claims](https://oidc.cirrus-ci.com/.well-known/openid-configuration).

```suggestion
| Google Cloud Build | yes          | yes             |                                                                                                                                                     |
| Cirrus CI.                  | yes          | yes             |                                                                                                                                                     |
```

Author: @fkorotkov love it! The next step would be adding support for it in Fulcio: https://github.com/sigstore/fulciox - I think some groundwork might need to be done here to extend the signing certificate. Here's an open issue for adding GitLab support for reference: https://github.com/sigstore/fulcio/issues/243

The release attestation will contain the following claims, linking the build to the release:
- Package org/name
- Package version
- Registry signature
- Tarball shasum

### Logging additional events on the Rekor transparency log

The current proposal only suggests publishing two events to the public Rekor log: sign during build and release/publish.

Publishing a richer set of events to the log could help create a safe and independent history of sensitive events. This could provide the possibility of detection when the registry DB or blob bucket are compromised, because it would be out of sync with the transparency log.

There's interest from OpenSSF's Securing Software Repos WG to standardize on a set of events to easy interoperability between different package registries.

Possible events that could be recorded in the transparency log:

- User registration
- Sign during build
- Push
- Publish or release
- Yank or delete
- Owner added
- Owner removed
- Owner changed email address
- Owner enabled MFA
- Owner disabled MFA
- Endorsement by non-owner
- Ownership call
- Ownership request
- Ownership transfer
- Resurrection of name

## Prior Art
### Sigstore cosign CLI

> The [cosign](https://github.com/sigstore/cosign) CLI tool supports signing arbitrary blobs and could be used to sign the package tarball manually but requires significant setup from maintainers and offers no simple way for consumers to verify signatures unless you know you to query for the package signature on Rekor.

Reviewer 12: This is also less safe if we allow non-trusted builds to sign the tarball as the build could falsify some elements of its provenance more easily.

Reviewer 13: Yes! Using a CI system to get the non-falsifiable context (e.g. git ref/sha) is a very important step.

Reviewer 3: How is that not falsifiable? CI systems can run `git` commands and set the sha to whatever they want.

Author: > How is that not falsifiable? CI systems can run `git` commands and set the sha to whatever they want.

This is a good point. As I see it, there are two sides to provenance information. One part we can get from the CI/CD systems OIDC id token. The content of this can be verified against the CIs public key so can't be falsified by some code running in CI/CD unless the CI provide is compromised. 

The id token will tell us which repo the action or CI job was triggered in but not necessarily which repo the build actually ran in. The information in the id token allows us to link back to the build logs for example (unless of course the repo has been deleted, which is still a problem here).

The other side is which repo the build was actually running in. This is slightly harder to lock down. This is why we're proposing a "trusted builder" that's a reusable workflow maintained by npm that can be used to build and publish the package. Using this workflow is recommended but optional as it might not be feasible to support every kind of publish workflow.

The trusted builder will execute the steps to checkout the repo from the information in the ID token. There's still a risk that the user supplied build steps run some code that modifies the checked out source repo. Locking this down further will be very hard until we have a way to run certain steps, e.g. `npm run build` without network access.

Reviewer 12: Maybe a more complete way of wording it would be end user non-falsifiable. Assuming you trust GitHub to do the right things then they should be the ones using `git` on your behalf. Compare that with the actual end user builds which can just run any arbitrary set of things.

Reviewer 18: It's also worth noting that the non-forgable part comes from the GitHub workflow run and OIDC identity attached to it. The idea is that a random person could not forge this identity but running the same commands on their laptop.

### SLSA Provenance Generator
Google is working on a language agnostic reusable Actions workflow that can generate build provenance claims using Sigstore: [slsa-github-generator](https://github.com/slsa-framework/slsa-github-generator)

It's technically possible to use something like this to produce verifiable build provenance when publishing an npm package but there's similarly to above, no easy way for package consumers to verify signatures in a standardized way when installing.

### SLSA Provenance Verifier

Implementation for verifying [SLSA Provenance](/Users/feelepxyz/src/github.com/npm/rfcs/accepted/0000-link-packages-to-source-and-build.md). It currently supports verifying provenance generated by the [SLSA generator for Go projects](https://github.com/slsa-framework/slsa-github-generator/blob/main/.github/workflows/builder_go_slsa3.yml).

## Unanswered questions

> - How could we support legacy packages that might not be maintained anymore?

Reviewer 12: Is rebuilding legacy packages off the table for NPM? I know that takes the original maintainer out of the loop. I just stat that because OpenSSF's Alpha Omega is open to potentially funding the safe building of packages.

Reviewer 8: I believe npm packages are considered immutable once published, and that the npm packages are not always reproducible.

A question: are there any circumstances where the source code might not be available? e.g. perhaps a binary could be embedded in an npm package without sharing the source?

Reviewer 3: Tons. Including packages whose repos are no longer available, or didn't contain a repo link at all.

Author: > Is rebuilding legacy packages off the table for NPM?

No but we won't be doing it as part of this project as its already a pretty huge initiative.

As I see it atm, the first step is laying the groundwork in npm for signing packages using some trusted identity, leveraging Sigstore as a vendor-neutral signing service. We'll start by supporting trusted build or workflow identities (CI/CD OIDC id tokens) for new packages.

Once we have this new capability in place we could investigate backfilling this information for legacy packages. For backfilled information the identity would need to be provided by npm instead of the CI/CD system but in principle it could leverage all the same functionality.

Reviewer 8: At the minimum it would represent a point in time snapshot of the current world. While not ideal, I think that does add value.

| CircleCI           | yes          | no              | Uses [workload ID](https://circleci.com/docs/2.0/openid-connect-tokens/) that includes a notion of “user ID” in the Fsubject claim. There is a [feature request to make this configurable](https://circleci.canny.io/cloud-feature-requests/p/customizable-audience-claim-in-oidc-tokens) |
| Google Cloud Build | yes          | yes             |                                                                                                                                                     |
| Cirrus CI.                  | yes          | yes             |                                                                                                                                                     |
| Bitbucket          | yes          | no              | Is it used for OSS?                                                                                                                                                    |

> | TravisCI           | no           | -               | Is it used for OSS?                                                                                                                                                    |

Author: ```suggestion
| TravisCI           | no           | -               |                                                                                                                                                    |
```

> | TravisCI           | no           | -               | Is it used for OSS?                                                                                                                                                    |

Author: Fixed, thanks for the context!

> | TravisCI           | no           | -               | Is it used for OSS?                                                                                                                                                    |

Reviewer 3: What they currently offer is a highly limited number of "free credits" a month, whereas the typical definition of "used for OSS" is "unlimited usage on public repos", so I'm not sure it qualifies without some kind of asterisk.

> | TravisCI           | no           | -               | Is it used for OSS?                                                                                                                                                    |

Reviewer 19: I agree it's worth adding an asterisk, because yes, it's a limited support due to those free credits situation. However in practice it works well, at least for my own OSS project, which is admittedly not very active.

> | TravisCI           | no           | -               | Is it used for OSS?                                                                                                                                                    |

Author: > 1 of those remaining 3 runs out of credits every month. It’s not sufficient.

This doesn't sound great! Will update this.

> | TravisCI           | no           | -               | Is it used for OSS?                                                                                                                                                    |

Reviewer 19: > and 1 of those remaining 3 runs out of credits every month
Yeah, that doesn't sound like a good deal indeed. Let's make that asterisk in bold.

## Non-goals
- Support for private/inner-source packages to begin with.
  - Our primary goal is to link packages hosted on the npm registry to the source and build it originated from, for this to be broadly useful, the source code and build need to be publicly available. We're committed to eventually supporting private use cases.
- Support signing from self-hosted CI/CD systems.
- Support signing from local machines or laptops.

>   - The initial goal does not support signing from your local laptop as our primary aim is to link the package to the source repo it was published from. There's no scalable way of making sure this source information isn't falsified when it comes from a local machine.

Reviewer 3: can you elaborate here on why a local machine that i can configure is different than a CI machine that i can configure?

Reviewer 12: It's a state space problem. A CI/CD system that is hosted immutably and can be verified/audited will be simpler to reason about compared with a laptop that usually has lots of arbitrary stuff installed.

Reviewer 3: @feelepxyz the possibility is only there if the community freely decides to make that move - and it’s not a free decision unless the existing mechanisms are equally privileged.

Reviewer: > Related to this, my hope is that all public packages are eventually published from open, auditable and automated systems with all the other benefits this brings like ephemeral environments and the ability to prove what went into a build. Maybe I'm deluding myself on the possibility of this though? 🤔

I think there are two reasons why this is challenging:

- exploration - whether from someone just learning JavaScript or from someone just exploring a new solution to a problem they have, this adds barriers and weight.
- tooling - currently, npm simply does not provide the necessary tools that make such automated publishing safe and reliable for anyone operating past a certain scale or who requires a certain level of trust.

I agree that things could be dramatically easier. Unfortunately that does require substantial investments, and it is seemingly often very challenging for certain teams to actually secure the necessary resources to be able to accomplish the outcomes that are required to achieve that ease.

## Implementation details

There will be two different ways to include provenance information during publish: running the CLI directly on a supported CI/CD system or using the optional but recommended "trusted builder to publish.

- "Sigstore integration in the npm CLI"
  - The npm CLI will include the capability to extract provenance information out-of-the-box on supported CI/CD systems.
  - The provenance information will be extracted from the ENV and ID token provided by the CI/CD system when running `npm publish`.
- "Non-falsifiable provenance using a trusted builder"
  - A reusable workflow or "trusted builder" can be used to invoke the npm CLI, extracting provenance, running install, build and publish in isolated steps, preventing malicious dependencies or build script from tampering with the provenance information.
  - The trusted builder will for example checkout the repo that's in the ID token as well as extract ENV information before any other user supplied commands are executed, passing this information to `npm publish` for signing.

>   - Using the trusted builder is recommended but it might not support all possible publish workflows so will be optional.

Reviewer 3: Does this mean that i can provide provenance information when building and publishing on my local machine?

Author: Not in this current proposal 😞  using the trusted builder will be optional when publishing from CI/CD.

I'm not opposed to supporting "author/maintainer" based identities for signing in future, for example using an email based identity or your npm account to sign the build. This would then allow you to publish from anywhere and bind a particular package to the maintainer.

This would be more like an extension to this proposal. I'd like to focus on build or workflow identities on CI/CD systems as a first step though to keep things manageable.
